// smart-trade-networks/lib/controllers/tokens/erc20Controller.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _typeof = require("@babel/runtime/helpers/typeof");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ERC20Controller = void 0;

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _ethers = require("ethers");

var _prefabContractFactory = require("../../chain/prefabContractFactory");

var artifact = _interopRequireWildcard(require("../../ethereum/abi/ERC20.json"));

var _encodeCall = require("../../ethereum/encodeCall");

var _ownableController = require("../helpers/ownableController");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2["default"])(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2["default"])(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2["default"])(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

var ERC20Controller = /*#__PURE__*/function (_OwnableController) {
  (0, _inherits2["default"])(ERC20Controller, _OwnableController);

  var _super = _createSuper(ERC20Controller);

  function ERC20Controller(erc20Address, signerOrProvider) {
    var _this;

    (0, _classCallCheck2["default"])(this, ERC20Controller);
    _this = _super.call(this, erc20Address, signerOrProvider);
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "_erc20Address", void 0);
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "_erc20Contract", void 0);
    _this._erc20Address = erc20Address;

    if (signerOrProvider instanceof _ethers.Signer || signerOrProvider instanceof _ethers.ethers.providers.JsonRpcSigner) {
      _this._signer = signerOrProvider;
    } else {
      var web3Wrapper = new _ethers.ethers.providers.Web3Provider(signerOrProvider);
      _this._signer = web3Wrapper.getSigner();
    }

    return _this;
  }

  (0, _createClass2["default"])(ERC20Controller, [{
    key: "_getERC20Contract",
    value: function () {
      var _getERC20Contract2 = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee() {
        var contract;
        return _regenerator["default"].wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                if (!this._erc20Contract) {
                  _context.next = 2;
                  break;
                }

                return _context.abrupt("return", this._erc20Contract);

              case 2:
                _context.prev = 2;
                _context.next = 5;
                return (0, _prefabContractFactory.getERC20Contract)(this._erc20Address, this._signer);

              case 5:
                contract = _context.sent;
                this._erc20Contract = contract;
                return _context.abrupt("return", contract);

              case 10:
                _context.prev = 10;
                _context.t0 = _context["catch"](2);
                throw Error("Failed to get ERC20 contract: ".concat(_context.t0));

              case 13:
              case "end":
                return _context.stop();
            }
          }
        }, _callee, this, [[2, 10]]);
      }));

      function _getERC20Contract() {
        return _getERC20Contract2.apply(this, arguments);
      }

      return _getERC20Contract;
    }()
    /**Getters */

  }, {
    key: "getContract",
    value: function () {
      var _getContract = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee2() {
        var erc20Contract;
        return _regenerator["default"].wrap(function _callee2$(_context2) {
          while (1) {
            switch (_context2.prev = _context2.next) {
              case 0:
                _context2.next = 2;
                return this._getERC20Contract();

              case 2:
                erc20Contract = _context2.sent;
                return _context2.abrupt("return", erc20Contract);

              case 4:
              case "end":
                return _context2.stop();
            }
          }
        }, _callee2, this);
      }));

      function getContract() {
        return _getContract.apply(this, arguments);
      }

      return getContract;
    }()
  }, {
    key: "balanceOf",
    value: function () {
      var _balanceOf = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee3(address) {
        var contract;
        return _regenerator["default"].wrap(function _callee3$(_context3) {
          while (1) {
            switch (_context3.prev = _context3.next) {
              case 0:
                _context3.next = 2;
                return this._getERC20Contract();

              case 2:
                contract = _context3.sent;
                return _context3.abrupt("return", contract.balanceOf(address));

              case 4:
              case "end":
                return _context3.stop();
            }
          }
        }, _callee3, this);
      }));

      function balanceOf(_x) {
        return _balanceOf.apply(this, arguments);
      }

      return balanceOf;
    }()
  }, {
    key: "getName",
    value: function () {
      var _getName = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee4() {
        var contract;
        return _regenerator["default"].wrap(function _callee4$(_context4) {
          while (1) {
            switch (_context4.prev = _context4.next) {
              case 0:
                _context4.next = 2;
                return this._getERC20Contract();

              case 2:
                contract = _context4.sent;
                return _context4.abrupt("return", contract.name());

              case 4:
              case "end":
                return _context4.stop();
            }
          }
        }, _callee4, this);
      }));

      function getName() {
        return _getName.apply(this, arguments);
      }

      return getName;
    }()
  }, {
    key: "getSymbol",
    value: function () {
      var _getSymbol = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee5() {
        var contract;
        return _regenerator["default"].wrap(function _callee5$(_context5) {
          while (1) {
            switch (_context5.prev = _context5.next) {
              case 0:
                _context5.next = 2;
                return this._getERC20Contract();

              case 2:
                contract = _context5.sent;
                return _context5.abrupt("return", contract.symbol());

              case 4:
              case "end":
                return _context5.stop();
            }
          }
        }, _callee5, this);
      }));

      function getSymbol() {
        return _getSymbol.apply(this, arguments);
      }

      return getSymbol;
    }()
  }, {
    key: "getTotalSupply",
    value: function () {
      var _getTotalSupply = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee6() {
        var contract;
        return _regenerator["default"].wrap(function _callee6$(_context6) {
          while (1) {
            switch (_context6.prev = _context6.next) {
              case 0:
                _context6.next = 2;
                return this._getERC20Contract();

              case 2:
                contract = _context6.sent;
                return _context6.abrupt("return", contract.totalSupply());

              case 4:
              case "end":
                return _context6.stop();
            }
          }
        }, _callee6, this);
      }));

      function getTotalSupply() {
        return _getTotalSupply.apply(this, arguments);
      }

      return getTotalSupply;
    }()
  }, {
    key: "getDecimals",
    value: function () {
      var _getDecimals = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee7() {
        var contract;
        return _regenerator["default"].wrap(function _callee7$(_context7) {
          while (1) {
            switch (_context7.prev = _context7.next) {
              case 0:
                _context7.next = 2;
                return this._getERC20Contract();

              case 2:
                contract = _context7.sent;
                return _context7.abrupt("return", contract.decimals());

              case 4:
              case "end":
                return _context7.stop();
            }
          }
        }, _callee7, this);
      }));

      function getDecimals() {
        return _getDecimals.apply(this, arguments);
      }

      return getDecimals;
    }()
  }, {
    key: "getAllowance",
    value: function () {
      var _getAllowance = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee8(owner, spender) {
        var contract;
        return _regenerator["default"].wrap(function _callee8$(_context8) {
          while (1) {
            switch (_context8.prev = _context8.next) {
              case 0:
                _context8.next = 2;
                return this._getERC20Contract();

              case 2:
                contract = _context8.sent;
                return _context8.abrupt("return", contract.allowance(owner, spender));

              case 4:
              case "end":
                return _context8.stop();
            }
          }
        }, _callee8, this);
      }));

      function getAllowance(_x2, _x3) {
        return _getAllowance.apply(this, arguments);
      }

      return getAllowance;
    }()
    /** Setters */

  }, {
    key: "transfer",
    value: function () {
      var _transfer = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee9(recipient, amount) {
        var contract;
        return _regenerator["default"].wrap(function _callee9$(_context9) {
          while (1) {
            switch (_context9.prev = _context9.next) {
              case 0:
                _context9.next = 2;
                return this._getERC20Contract();

              case 2:
                contract = _context9.sent;
                return _context9.abrupt("return", contract.functions.transfer(recipient, amount));

              case 4:
              case "end":
                return _context9.stop();
            }
          }
        }, _callee9, this);
      }));

      function transfer(_x4, _x5) {
        return _transfer.apply(this, arguments);
      }

      return transfer;
    }()
  }, {
    key: "approve",
    value: function () {
      var _approve = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee10(spender, amount) {
        var contract;
        return _regenerator["default"].wrap(function _callee10$(_context10) {
          while (1) {
            switch (_context10.prev = _context10.next) {
              case 0:
                _context10.next = 2;
                return this._getERC20Contract();

              case 2:
                contract = _context10.sent;
                return _context10.abrupt("return", contract.functions.approve(spender, amount));

              case 4:
              case "end":
                return _context10.stop();
            }
          }
        }, _callee10, this);
      }));

      function approve(_x6, _x7) {
        return _approve.apply(this, arguments);
      }

      return approve;
    }()
  }, {
    key: "transferFrom",
    value: function () {
      var _transferFrom = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee11(sender, recipient, amount) {
        var contract;
        return _regenerator["default"].wrap(function _callee11$(_context11) {
          while (1) {
            switch (_context11.prev = _context11.next) {
              case 0:
                _context11.next = 2;
                return this._getERC20Contract();

              case 2:
                contract = _context11.sent;
                return _context11.abrupt("return", contract.functions.transferFrom(sender, recipient, amount));

              case 4:
              case "end":
                return _context11.stop();
            }
          }
        }, _callee11, this);
      }));

      function transferFrom(_x8, _x9, _x10) {
        return _transferFrom.apply(this, arguments);
      }

      return transferFrom;
    }()
  }, {
    key: "increaseAllowance",
    value: function () {
      var _increaseAllowance = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee12(spender, addedValue) {
        var contract;
        return _regenerator["default"].wrap(function _callee12$(_context12) {
          while (1) {
            switch (_context12.prev = _context12.next) {
              case 0:
                _context12.next = 2;
                return this._getERC20Contract();

              case 2:
                contract = _context12.sent;
                return _context12.abrupt("return", contract.functions.increaseAllowance(spender, addedValue));

              case 4:
              case "end":
                return _context12.stop();
            }
          }
        }, _callee12, this);
      }));

      function increaseAllowance(_x11, _x12) {
        return _increaseAllowance.apply(this, arguments);
      }

      return increaseAllowance;
    }()
  }, {
    key: "decreaseAllowance",
    value: function () {
      var _decreaseAllowance = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee13(spender, subtractedValue) {
        var contract;
        return _regenerator["default"].wrap(function _callee13$(_context13) {
          while (1) {
            switch (_context13.prev = _context13.next) {
              case 0:
                _context13.next = 2;
                return this._getERC20Contract();

              case 2:
                contract = _context13.sent;
                return _context13.abrupt("return", contract.functions.decreaseAllowance(spender, subtractedValue));

              case 4:
              case "end":
                return _context13.stop();
            }
          }
        }, _callee13, this);
      }));

      function decreaseAllowance(_x13, _x14) {
        return _decreaseAllowance.apply(this, arguments);
      }

      return decreaseAllowance;
    }()
    /** Helpers */

  }, {
    key: "onEvent",
    value: function () {
      var _onEvent = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee14(filter, onContractevent) {
        var contract;
        return _regenerator["default"].wrap(function _callee14$(_context14) {
          while (1) {
            switch (_context14.prev = _context14.next) {
              case 0:
                _context14.next = 2;
                return this._getERC20Contract();

              case 2:
                contract = _context14.sent;
                contract.once(filter, function (from, to, value, event) {
                  onContractevent({
                    from: from,
                    to: to,
                    value: value,
                    event: event
                  });
                });

              case 4:
              case "end":
                return _context14.stop();
            }
          }
        }, _callee14, this);
      }));

      function onEvent(_x15, _x16) {
        return _onEvent.apply(this, arguments);
      }

      return onEvent;
    }()
  }], [{
    key: "encodeCall",
    value: function encodeCall(functionName, args) {
      return (0, _encodeCall.encode)(artifact, functionName, args);
    }
  }, {
    key: "decodeCall",
    value: function decodeCall(data) {
      return (0, _encodeCall.decode)(artifact, data);
    }
  }]);
  return ERC20Controller;
}(_ownableController.OwnableController);

exports.ERC20Controller = ERC20Controller;
