// smart-trade-networks/lib/controllers/tokens/erc721Controller.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ERC721Controller = void 0;

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _ethers = require("ethers");

var _prefabContractFactory = require("../../chain/prefabContractFactory");

var ERC721Controller = /*#__PURE__*/function () {
  function ERC721Controller(erc721Address, signerOrProvider) {
    (0, _classCallCheck2["default"])(this, ERC721Controller);
    (0, _defineProperty2["default"])(this, "_signer", void 0);
    (0, _defineProperty2["default"])(this, "_erc721Address", void 0);
    (0, _defineProperty2["default"])(this, "_erc721Contract", void 0);
    this._erc721Address = erc721Address;

    if (signerOrProvider instanceof _ethers.Signer || signerOrProvider instanceof _ethers.ethers.providers.JsonRpcSigner) {
      this._signer = signerOrProvider;
    } else {
      var web3Wrapper = new _ethers.ethers.providers.Web3Provider(signerOrProvider);
      this._signer = web3Wrapper.getSigner();
    }
  }

  (0, _createClass2["default"])(ERC721Controller, [{
    key: "_getERC721Contract",
    value: function () {
      var _getERC721Contract2 = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee() {
        var contract;
        return _regenerator["default"].wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                if (!this._erc721Contract) {
                  _context.next = 2;
                  break;
                }

                return _context.abrupt("return", this._erc721Contract);

              case 2:
                _context.prev = 2;
                _context.next = 5;
                return (0, _prefabContractFactory.getERC721Contract)(this._erc721Address, this._signer);

              case 5:
                contract = _context.sent;
                this._erc721Contract = contract;
                return _context.abrupt("return", contract);

              case 10:
                _context.prev = 10;
                _context.t0 = _context["catch"](2);
                throw Error("Failed to get ERC721 contract: ".concat(_context.t0));

              case 13:
              case "end":
                return _context.stop();
            }
          }
        }, _callee, this, [[2, 10]]);
      }));

      function _getERC721Contract() {
        return _getERC721Contract2.apply(this, arguments);
      }

      return _getERC721Contract;
    }()
    /** Getters */

  }, {
    key: "balanceOf",
    value: function () {
      var _balanceOf = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee2(user) {
        var erc721Contract;
        return _regenerator["default"].wrap(function _callee2$(_context2) {
          while (1) {
            switch (_context2.prev = _context2.next) {
              case 0:
                _context2.next = 2;
                return this._getERC721Contract();

              case 2:
                erc721Contract = _context2.sent;
                return _context2.abrupt("return", erc721Contract.balanceOf(user));

              case 4:
              case "end":
                return _context2.stop();
            }
          }
        }, _callee2, this);
      }));

      function balanceOf(_x) {
        return _balanceOf.apply(this, arguments);
      }

      return balanceOf;
    }()
  }, {
    key: "getApproved",
    value: function () {
      var _getApproved = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee3(tokenId) {
        var erc721Contract;
        return _regenerator["default"].wrap(function _callee3$(_context3) {
          while (1) {
            switch (_context3.prev = _context3.next) {
              case 0:
                _context3.next = 2;
                return this._getERC721Contract();

              case 2:
                erc721Contract = _context3.sent;
                return _context3.abrupt("return", erc721Contract.getApproved(tokenId));

              case 4:
              case "end":
                return _context3.stop();
            }
          }
        }, _callee3, this);
      }));

      function getApproved(_x2) {
        return _getApproved.apply(this, arguments);
      }

      return getApproved;
    }()
  }, {
    key: "isApprovedForAll",
    value: function () {
      var _isApprovedForAll = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee4(owner, operator) {
        var erc721Contract;
        return _regenerator["default"].wrap(function _callee4$(_context4) {
          while (1) {
            switch (_context4.prev = _context4.next) {
              case 0:
                _context4.next = 2;
                return this._getERC721Contract();

              case 2:
                erc721Contract = _context4.sent;
                return _context4.abrupt("return", erc721Contract.isApprovedForAll(owner, operator));

              case 4:
              case "end":
                return _context4.stop();
            }
          }
        }, _callee4, this);
      }));

      function isApprovedForAll(_x3, _x4) {
        return _isApprovedForAll.apply(this, arguments);
      }

      return isApprovedForAll;
    }()
  }, {
    key: "exists",
    value: function () {
      var _exists = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee5(tokenId) {
        var erc721Contract;
        return _regenerator["default"].wrap(function _callee5$(_context5) {
          while (1) {
            switch (_context5.prev = _context5.next) {
              case 0:
                _context5.next = 2;
                return this._getERC721Contract();

              case 2:
                erc721Contract = _context5.sent;
                return _context5.abrupt("return", erc721Contract.exists(tokenId));

              case 4:
              case "end":
                return _context5.stop();
            }
          }
        }, _callee5, this);
      }));

      function exists(_x5) {
        return _exists.apply(this, arguments);
      }

      return exists;
    }()
  }, {
    key: "getTotalTokens",
    value: function () {
      var _getTotalTokens = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee6() {
        var erc721Contract;
        return _regenerator["default"].wrap(function _callee6$(_context6) {
          while (1) {
            switch (_context6.prev = _context6.next) {
              case 0:
                _context6.next = 2;
                return this._getERC721Contract();

              case 2:
                erc721Contract = _context6.sent;
                return _context6.abrupt("return", erc721Contract.getTotalTokens());

              case 4:
              case "end":
                return _context6.stop();
            }
          }
        }, _callee6, this);
      }));

      function getTotalTokens() {
        return _getTotalTokens.apply(this, arguments);
      }

      return getTotalTokens;
    }()
  }, {
    key: "getOwnerOf",
    value: function () {
      var _getOwnerOf = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee7(tokenId) {
        var erc721Contract;
        return _regenerator["default"].wrap(function _callee7$(_context7) {
          while (1) {
            switch (_context7.prev = _context7.next) {
              case 0:
                _context7.next = 2;
                return this._getERC721Contract();

              case 2:
                erc721Contract = _context7.sent;
                return _context7.abrupt("return", erc721Contract.ownerOf(tokenId));

              case 4:
              case "end":
                return _context7.stop();
            }
          }
        }, _callee7, this);
      }));

      function getOwnerOf(_x6) {
        return _getOwnerOf.apply(this, arguments);
      }

      return getOwnerOf;
    }()
  }, {
    key: "getOwner",
    value: function () {
      var _getOwner = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee8() {
        var erc721Contract;
        return _regenerator["default"].wrap(function _callee8$(_context8) {
          while (1) {
            switch (_context8.prev = _context8.next) {
              case 0:
                _context8.next = 2;
                return this._getERC721Contract();

              case 2:
                erc721Contract = _context8.sent;
                return _context8.abrupt("return", erc721Contract.owner());

              case 4:
              case "end":
                return _context8.stop();
            }
          }
        }, _callee8, this);
      }));

      function getOwner() {
        return _getOwner.apply(this, arguments);
      }

      return getOwner;
    }()
    /** Version 1_1 */

  }, {
    key: "getGroupingAddress",
    value: function () {
      var _getGroupingAddress = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee9(tokenId) {
        var erc721Contract;
        return _regenerator["default"].wrap(function _callee9$(_context9) {
          while (1) {
            switch (_context9.prev = _context9.next) {
              case 0:
                _context9.next = 2;
                return this._getERC721Contract();

              case 2:
                erc721Contract = _context9.sent;
                return _context9.abrupt("return", erc721Contract.getGroupingAddress(tokenId));

              case 4:
              case "end":
                return _context9.stop();
            }
          }
        }, _callee9, this);
      }));

      function getGroupingAddress(_x7) {
        return _getGroupingAddress.apply(this, arguments);
      }

      return getGroupingAddress;
    }()
    /**Setters */

  }, {
    key: "approve",
    value: function () {
      var _approve = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee10(to, tokenId) {
        var erc721Contract;
        return _regenerator["default"].wrap(function _callee10$(_context10) {
          while (1) {
            switch (_context10.prev = _context10.next) {
              case 0:
                _context10.next = 2;
                return this._getERC721Contract();

              case 2:
                erc721Contract = _context10.sent;
                return _context10.abrupt("return", erc721Contract.functions.approve(to, tokenId));

              case 4:
              case "end":
                return _context10.stop();
            }
          }
        }, _callee10, this);
      }));

      function approve(_x8, _x9) {
        return _approve.apply(this, arguments);
      }

      return approve;
    }()
  }, {
    key: "setApprovalForAll",
    value: function () {
      var _setApprovalForAll = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee11(to, approved) {
        var erc721Contract;
        return _regenerator["default"].wrap(function _callee11$(_context11) {
          while (1) {
            switch (_context11.prev = _context11.next) {
              case 0:
                _context11.next = 2;
                return this._getERC721Contract();

              case 2:
                erc721Contract = _context11.sent;
                return _context11.abrupt("return", erc721Contract.functions.setApprovalForAll(to, approved));

              case 4:
              case "end":
                return _context11.stop();
            }
          }
        }, _callee11, this);
      }));

      function setApprovalForAll(_x10, _x11) {
        return _setApprovalForAll.apply(this, arguments);
      }

      return setApprovalForAll;
    }()
  }, {
    key: "transferFrom",
    value: function () {
      var _transferFrom = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee12(from, to, tokenId) {
        var erc721Contract;
        return _regenerator["default"].wrap(function _callee12$(_context12) {
          while (1) {
            switch (_context12.prev = _context12.next) {
              case 0:
                _context12.next = 2;
                return this._getERC721Contract();

              case 2:
                erc721Contract = _context12.sent;
                return _context12.abrupt("return", erc721Contract.functions.transferFrom(from, to, tokenId));

              case 4:
              case "end":
                return _context12.stop();
            }
          }
        }, _callee12, this);
      }));

      function transferFrom(_x12, _x13, _x14) {
        return _transferFrom.apply(this, arguments);
      }

      return transferFrom;
    }()
  }]);
  return ERC721Controller;
}();

exports.ERC721Controller = ERC721Controller;
