// smart-trade-networks/lib/controllers/tokens/stn721Controller.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.STN721Controller = void 0;

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _ethers = require("ethers");

var _prefabContractFactory = require("../../chain/prefabContractFactory");

var _deploy = require("../../ethereum/deploy/deploy");

var _erc721Controller = require("./erc721Controller");

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2["default"])(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2["default"])(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2["default"])(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

var STN721Controller = /*#__PURE__*/function (_ERC721Controller) {
  (0, _inherits2["default"])(STN721Controller, _ERC721Controller);

  var _super = _createSuper(STN721Controller);

  function STN721Controller(stn721Address, signerOrProvider) {
    var _this;

    (0, _classCallCheck2["default"])(this, STN721Controller);
    _this = _super.call(this, stn721Address, signerOrProvider);
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "_stn721Address", void 0);
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "_stn721Contract", void 0);
    _this._stn721Address = stn721Address;

    if (signerOrProvider instanceof _ethers.Signer || signerOrProvider instanceof _ethers.ethers.providers.JsonRpcSigner) {
      _this._signer = signerOrProvider;
    } else {
      var web3Wrapper = new _ethers.ethers.providers.Web3Provider(signerOrProvider);
      _this._signer = web3Wrapper.getSigner();
    }

    return _this;
  }

  (0, _createClass2["default"])(STN721Controller, [{
    key: "_getSTN721Contract",
    value: function () {
      var _getSTN721Contract2 = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee() {
        var contract;
        return _regenerator["default"].wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                if (!this._stn721Contract) {
                  _context.next = 2;
                  break;
                }

                return _context.abrupt("return", this._stn721Contract);

              case 2:
                _context.prev = 2;
                _context.next = 5;
                return (0, _prefabContractFactory.getSTN721Contract)(this._stn721Address, this._signer);

              case 5:
                contract = _context.sent;
                this._stn721Contract = contract;
                return _context.abrupt("return", contract);

              case 10:
                _context.prev = 10;
                _context.t0 = _context["catch"](2);
                throw Error("Failed to get STN721 contract: ".concat(_context.t0));

              case 13:
              case "end":
                return _context.stop();
            }
          }
        }, _callee, this, [[2, 10]]);
      }));

      function _getSTN721Contract() {
        return _getSTN721Contract2.apply(this, arguments);
      }

      return _getSTN721Contract;
    }()
  }, {
    key: "getContract",
    value:
    /**Getters */
    function () {
      var _getContract = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee2() {
        var contract;
        return _regenerator["default"].wrap(function _callee2$(_context2) {
          while (1) {
            switch (_context2.prev = _context2.next) {
              case 0:
                _context2.next = 2;
                return this._getSTN721Contract();

              case 2:
                contract = _context2.sent;
                return _context2.abrupt("return", contract);

              case 4:
              case "end":
                return _context2.stop();
            }
          }
        }, _callee2, this);
      }));

      function getContract() {
        return _getContract.apply(this, arguments);
      }

      return getContract;
    }()
  }, {
    key: "getParentAddress",
    value: function () {
      var _getParentAddress = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee3() {
        var contract;
        return _regenerator["default"].wrap(function _callee3$(_context3) {
          while (1) {
            switch (_context3.prev = _context3.next) {
              case 0:
                _context3.next = 2;
                return this._getSTN721Contract();

              case 2:
                contract = _context3.sent;
                return _context3.abrupt("return", contract.getParentAddress());

              case 4:
              case "end":
                return _context3.stop();
            }
          }
        }, _callee3, this);
      }));

      function getParentAddress() {
        return _getParentAddress.apply(this, arguments);
      }

      return getParentAddress;
    }()
  }, {
    key: "getParentId",
    value: function () {
      var _getParentId = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee4() {
        var contract;
        return _regenerator["default"].wrap(function _callee4$(_context4) {
          while (1) {
            switch (_context4.prev = _context4.next) {
              case 0:
                _context4.next = 2;
                return this._getSTN721Contract();

              case 2:
                contract = _context4.sent;
                return _context4.abrupt("return", contract.getParentId());

              case 4:
              case "end":
                return _context4.stop();
            }
          }
        }, _callee4, this);
      }));

      function getParentId() {
        return _getParentId.apply(this, arguments);
      }

      return getParentId;
    }()
  }, {
    key: "isDivisible",
    value: function () {
      var _isDivisible = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee5(tokenId) {
        var contract;
        return _regenerator["default"].wrap(function _callee5$(_context5) {
          while (1) {
            switch (_context5.prev = _context5.next) {
              case 0:
                _context5.next = 2;
                return this._getSTN721Contract();

              case 2:
                contract = _context5.sent;
                return _context5.abrupt("return", contract.isDivisible(tokenId));

              case 4:
              case "end":
                return _context5.stop();
            }
          }
        }, _callee5, this);
      }));

      function isDivisible(_x) {
        return _isDivisible.apply(this, arguments);
      }

      return isDivisible;
    }()
  }, {
    key: "getChildIndexById",
    value: function () {
      var _getChildIndexById = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee6(tokenId) {
        var contract;
        return _regenerator["default"].wrap(function _callee6$(_context6) {
          while (1) {
            switch (_context6.prev = _context6.next) {
              case 0:
                _context6.next = 2;
                return this._getSTN721Contract();

              case 2:
                contract = _context6.sent;
                return _context6.abrupt("return", contract.getChildIndexById(tokenId));

              case 4:
              case "end":
                return _context6.stop();
            }
          }
        }, _callee6, this);
      }));

      function getChildIndexById(_x2) {
        return _getChildIndexById.apply(this, arguments);
      }

      return getChildIndexById;
    }()
  }, {
    key: "getChildIndexByAddress",
    value: function () {
      var _getChildIndexByAddress = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee7(stn721) {
        var contract;
        return _regenerator["default"].wrap(function _callee7$(_context7) {
          while (1) {
            switch (_context7.prev = _context7.next) {
              case 0:
                _context7.next = 2;
                return this._getSTN721Contract();

              case 2:
                contract = _context7.sent;
                return _context7.abrupt("return", contract.getChildIndexByAddress(stn721));

              case 4:
              case "end":
                return _context7.stop();
            }
          }
        }, _callee7, this);
      }));

      function getChildIndexByAddress(_x3) {
        return _getChildIndexByAddress.apply(this, arguments);
      }

      return getChildIndexByAddress;
    }()
    /** Setters */

  }, {
    key: "attachChild",
    value: function () {
      var _attachChild = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee8(stn721Child, tokenId) {
        var contract;
        return _regenerator["default"].wrap(function _callee8$(_context8) {
          while (1) {
            switch (_context8.prev = _context8.next) {
              case 0:
                _context8.next = 2;
                return this._getSTN721Contract();

              case 2:
                contract = _context8.sent;
                return _context8.abrupt("return", contract.functions.attachChild(stn721Child, tokenId));

              case 4:
              case "end":
                return _context8.stop();
            }
          }
        }, _callee8, this);
      }));

      function attachChild(_x4, _x5) {
        return _attachChild.apply(this, arguments);
      }

      return attachChild;
    }()
  }, {
    key: "mint",
    value: function () {
      var _mint = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee9(divisible, tokenIds, receiver) {
        var contract;
        return _regenerator["default"].wrap(function _callee9$(_context9) {
          while (1) {
            switch (_context9.prev = _context9.next) {
              case 0:
                _context9.next = 2;
                return this._getSTN721Contract();

              case 2:
                contract = _context9.sent;
                return _context9.abrupt("return", contract.functions.mint(divisible, tokenIds, receiver));

              case 4:
              case "end":
                return _context9.stop();
            }
          }
        }, _callee9, this);
      }));

      function mint(_x6, _x7, _x8) {
        return _mint.apply(this, arguments);
      }

      return mint;
    }()
  }], [{
    key: "deploySTN721Contract",
    value: function () {
      var _deploySTN721Contract = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee10(signer, parentAddress, parentId, name, symbol) {
        return _regenerator["default"].wrap(function _callee10$(_context10) {
          while (1) {
            switch (_context10.prev = _context10.next) {
              case 0:
                return _context10.abrupt("return", (0, _deploy.deploySTN721)(signer, parentAddress, parentId, name, symbol));

              case 1:
              case "end":
                return _context10.stop();
            }
          }
        }, _callee10);
      }));

      function deploySTN721Contract(_x9, _x10, _x11, _x12, _x13) {
        return _deploySTN721Contract.apply(this, arguments);
      }

      return deploySTN721Contract;
    }()
  }]);
  return STN721Controller;
}(_erc721Controller.ERC721Controller);

exports.STN721Controller = STN721Controller;
