// smart-trade-networks/lib/controllers/rea/companiesFactoryController.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _typeof = require("@babel/runtime/helpers/typeof");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CompaniesFactoryController = void 0;

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _ethers = require("ethers");

var _prefabContractFactory = require("../../chain/prefabContractFactory");

var _deploy = require("../../ethereum/deploy/deploy");

var _encodeCall = require("../../ethereum/encodeCall");

var artifact = _interopRequireWildcard(require("../../ethereum/abi/CompaniesFactory.json"));

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

var CompaniesFactoryController = /*#__PURE__*/function () {
  function CompaniesFactoryController(companiesFactoryAddress, signerOrProvider) {
    (0, _classCallCheck2["default"])(this, CompaniesFactoryController);
    (0, _defineProperty2["default"])(this, "_companiesFactoryAddress", void 0);
    (0, _defineProperty2["default"])(this, "_signer", void 0);
    (0, _defineProperty2["default"])(this, "_factoryContract", void 0);
    this._companiesFactoryAddress = companiesFactoryAddress;

    if (signerOrProvider instanceof _ethers.Signer || signerOrProvider instanceof _ethers.ethers.providers.JsonRpcSigner) {
      this._signer = signerOrProvider;
    } else {
      var web3Wrapper = new _ethers.ethers.providers.Web3Provider(signerOrProvider);
      this._signer = web3Wrapper.getSigner();
    }
  }

  (0, _createClass2["default"])(CompaniesFactoryController, [{
    key: "_getCompaniesFactoryContract",
    value: function () {
      var _getCompaniesFactoryContract2 = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee() {
        var contract;
        return _regenerator["default"].wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                if (!this._factoryContract) {
                  _context.next = 2;
                  break;
                }

                return _context.abrupt("return", this._factoryContract);

              case 2:
                _context.prev = 2;
                _context.next = 5;
                return (0, _prefabContractFactory.getCompaniesFactoryContract)(this._companiesFactoryAddress, this._signer);

              case 5:
                contract = _context.sent;
                this._factoryContract = contract;
                return _context.abrupt("return", contract);

              case 10:
                _context.prev = 10;
                _context.t0 = _context["catch"](2);
                throw Error("Failed to get CompaniesFactory contract: ".concat(_context.t0));

              case 13:
              case "end":
                return _context.stop();
            }
          }
        }, _callee, this, [[2, 10]]);
      }));

      function _getCompaniesFactoryContract() {
        return _getCompaniesFactoryContract2.apply(this, arguments);
      }

      return _getCompaniesFactoryContract;
    }()
  }, {
    key: "getContract",
    value:
    /**Getters */
    function () {
      var _getContract = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee2() {
        var contract;
        return _regenerator["default"].wrap(function _callee2$(_context2) {
          while (1) {
            switch (_context2.prev = _context2.next) {
              case 0:
                _context2.next = 2;
                return this._getCompaniesFactoryContract();

              case 2:
                contract = _context2.sent;
                return _context2.abrupt("return", contract);

              case 4:
              case "end":
                return _context2.stop();
            }
          }
        }, _callee2, this);
      }));

      function getContract() {
        return _getContract.apply(this, arguments);
      }

      return getContract;
    }()
  }, {
    key: "getCompanies",
    value: function () {
      var _getCompanies = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee3() {
        var contract;
        return _regenerator["default"].wrap(function _callee3$(_context3) {
          while (1) {
            switch (_context3.prev = _context3.next) {
              case 0:
                _context3.next = 2;
                return this._getCompaniesFactoryContract();

              case 2:
                contract = _context3.sent;
                return _context3.abrupt("return", contract.getCompanies());

              case 4:
              case "end":
                return _context3.stop();
            }
          }
        }, _callee3, this);
      }));

      function getCompanies() {
        return _getCompanies.apply(this, arguments);
      }

      return getCompanies;
    }()
  }, {
    key: "isCompanyOwner",
    value: function () {
      var _isCompanyOwner = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee4(companyAddress, owner) {
        var contract;
        return _regenerator["default"].wrap(function _callee4$(_context4) {
          while (1) {
            switch (_context4.prev = _context4.next) {
              case 0:
                _context4.next = 2;
                return this._getCompaniesFactoryContract();

              case 2:
                contract = _context4.sent;
                return _context4.abrupt("return", contract.isCompanyOwner(companyAddress, owner));

              case 4:
              case "end":
                return _context4.stop();
            }
          }
        }, _callee4, this);
      }));

      function isCompanyOwner(_x, _x2) {
        return _isCompanyOwner.apply(this, arguments);
      }

      return isCompanyOwner;
    }()
  }, {
    key: "getCompaniesBySignatorie",
    value: function () {
      var _getCompaniesBySignatorie = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee5(signatorie) {
        var contract;
        return _regenerator["default"].wrap(function _callee5$(_context5) {
          while (1) {
            switch (_context5.prev = _context5.next) {
              case 0:
                _context5.next = 2;
                return this._getCompaniesFactoryContract();

              case 2:
                contract = _context5.sent;
                return _context5.abrupt("return", contract.getCompaniesBySignatorie(signatorie));

              case 4:
              case "end":
                return _context5.stop();
            }
          }
        }, _callee5, this);
      }));

      function getCompaniesBySignatorie(_x3) {
        return _getCompaniesBySignatorie.apply(this, arguments);
      }

      return getCompaniesBySignatorie;
    }()
  }, {
    key: "getSignersByCompany",
    value: function () {
      var _getSignersByCompany = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee6(companyAddress) {
        var contract;
        return _regenerator["default"].wrap(function _callee6$(_context6) {
          while (1) {
            switch (_context6.prev = _context6.next) {
              case 0:
                _context6.next = 2;
                return this._getCompaniesFactoryContract();

              case 2:
                contract = _context6.sent;
                return _context6.abrupt("return", contract.getSignersByCompany(companyAddress));

              case 4:
              case "end":
                return _context6.stop();
            }
          }
        }, _callee6, this);
      }));

      function getSignersByCompany(_x4) {
        return _getSignersByCompany.apply(this, arguments);
      }

      return getSignersByCompany;
    }()
    /** Setters */

  }, {
    key: "createCompany",
    value: function () {
      var _createCompany = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee7(owner, name, ipfsLink) {
        var contract;
        return _regenerator["default"].wrap(function _callee7$(_context7) {
          while (1) {
            switch (_context7.prev = _context7.next) {
              case 0:
                _context7.next = 2;
                return this._getCompaniesFactoryContract();

              case 2:
                contract = _context7.sent;
                return _context7.abrupt("return", contract.functions.createCompany(owner, name, ipfsLink));

              case 4:
              case "end":
                return _context7.stop();
            }
          }
        }, _callee7, this);
      }));

      function createCompany(_x5, _x6, _x7) {
        return _createCompany.apply(this, arguments);
      }

      return createCompany;
    }()
  }, {
    key: "removeCompany",
    value: function () {
      var _removeCompany = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee8(companyAddress) {
        var contract;
        return _regenerator["default"].wrap(function _callee8$(_context8) {
          while (1) {
            switch (_context8.prev = _context8.next) {
              case 0:
                _context8.next = 2;
                return this._getCompaniesFactoryContract();

              case 2:
                contract = _context8.sent;
                return _context8.abrupt("return", contract.functions.removeCompany(companyAddress));

              case 4:
              case "end":
                return _context8.stop();
            }
          }
        }, _callee8, this);
      }));

      function removeCompany(_x8) {
        return _removeCompany.apply(this, arguments);
      }

      return removeCompany;
    }()
  }, {
    key: "approveCompany",
    value: function () {
      var _approveCompany = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee9(companyAddress) {
        var contract;
        return _regenerator["default"].wrap(function _callee9$(_context9) {
          while (1) {
            switch (_context9.prev = _context9.next) {
              case 0:
                _context9.next = 2;
                return this._getCompaniesFactoryContract();

              case 2:
                contract = _context9.sent;
                return _context9.abrupt("return", contract.functions.approveCompany(companyAddress));

              case 4:
              case "end":
                return _context9.stop();
            }
          }
        }, _callee9, this);
      }));

      function approveCompany(_x9) {
        return _approveCompany.apply(this, arguments);
      }

      return approveCompany;
    }()
  }, {
    key: "addCompanySignatorie",
    value: function () {
      var _addCompanySignatorie = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee10(companyAddress, newSignatorie) {
        var contract;
        return _regenerator["default"].wrap(function _callee10$(_context10) {
          while (1) {
            switch (_context10.prev = _context10.next) {
              case 0:
                _context10.next = 2;
                return this._getCompaniesFactoryContract();

              case 2:
                contract = _context10.sent;
                return _context10.abrupt("return", contract.functions.addCompanySignatorie(companyAddress, newSignatorie));

              case 4:
              case "end":
                return _context10.stop();
            }
          }
        }, _callee10, this);
      }));

      function addCompanySignatorie(_x10, _x11) {
        return _addCompanySignatorie.apply(this, arguments);
      }

      return addCompanySignatorie;
    }()
  }, {
    key: "removeCompanySignatorie",
    value: function () {
      var _removeCompanySignatorie = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee11(companyAddress, signatorie) {
        var contract;
        return _regenerator["default"].wrap(function _callee11$(_context11) {
          while (1) {
            switch (_context11.prev = _context11.next) {
              case 0:
                _context11.next = 2;
                return this._getCompaniesFactoryContract();

              case 2:
                contract = _context11.sent;
                return _context11.abrupt("return", contract.functions.removeCompanySignatorie(companyAddress, signatorie));

              case 4:
              case "end":
                return _context11.stop();
            }
          }
        }, _callee11, this);
      }));

      function removeCompanySignatorie(_x12, _x13) {
        return _removeCompanySignatorie.apply(this, arguments);
      }

      return removeCompanySignatorie;
    }()
    /** Helpers */

  }], [{
    key: "deployCompaniesFactory",
    value: function () {
      var _deployCompaniesFactory = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee12(signer, owner) {
        return _regenerator["default"].wrap(function _callee12$(_context12) {
          while (1) {
            switch (_context12.prev = _context12.next) {
              case 0:
                return _context12.abrupt("return", (0, _deploy.deployCompaniesfactory)(signer, owner));

              case 1:
              case "end":
                return _context12.stop();
            }
          }
        }, _callee12);
      }));

      function deployCompaniesFactory(_x14, _x15) {
        return _deployCompaniesFactory.apply(this, arguments);
      }

      return deployCompaniesFactory;
    }()
  }, {
    key: "encodeCall",
    value: function encodeCall(functionName, args) {
      return (0, _encodeCall.encode)(artifact, functionName, args);
    }
  }, {
    key: "decodeCall",
    value: function decodeCall(data) {
      return (0, _encodeCall.decode)(artifact, data);
    }
  }]);
  return CompaniesFactoryController;
}();

exports.CompaniesFactoryController = CompaniesFactoryController;
