// smart-trade-networks/lib/controllers/rea/managerController.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ManagerController = void 0;

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _ethers = require("ethers");

var _prefabContractFactory = require("../../chain/prefabContractFactory");

var _deploy = require("../../ethereum/deploy/deploy");

var ManagerController = /*#__PURE__*/function () {
  function ManagerController(managerAddress, signerOrProvider) {
    (0, _classCallCheck2["default"])(this, ManagerController);
    (0, _defineProperty2["default"])(this, "_managerAddress", void 0);
    (0, _defineProperty2["default"])(this, "_signer", void 0);
    (0, _defineProperty2["default"])(this, "_managerContract", void 0);
    this._managerAddress = managerAddress;

    if (signerOrProvider instanceof _ethers.Signer || signerOrProvider instanceof _ethers.ethers.providers.JsonRpcSigner) {
      this._signer = signerOrProvider;
    } else {
      var web3Wrapper = new _ethers.ethers.providers.Web3Provider(signerOrProvider);
      this._signer = web3Wrapper.getSigner();
    }
  }

  (0, _createClass2["default"])(ManagerController, [{
    key: "_getManagerContract",
    value: function () {
      var _getManagerContract2 = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee() {
        var contract;
        return _regenerator["default"].wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                if (!this._managerContract) {
                  _context.next = 2;
                  break;
                }

                return _context.abrupt("return", this._managerContract);

              case 2:
                _context.prev = 2;
                _context.next = 5;
                return (0, _prefabContractFactory.getManagerContract)(this._managerAddress, this._signer);

              case 5:
                contract = _context.sent;
                this._managerContract = contract;
                return _context.abrupt("return", contract);

              case 10:
                _context.prev = 10;
                _context.t0 = _context["catch"](2);
                throw Error("Failed to get STN721Manager contract: ".concat(_context.t0));

              case 13:
              case "end":
                return _context.stop();
            }
          }
        }, _callee, this, [[2, 10]]);
      }));

      function _getManagerContract() {
        return _getManagerContract2.apply(this, arguments);
      }

      return _getManagerContract;
    }()
  }, {
    key: "getContract",
    value:
    /**Getters */
    function () {
      var _getContract = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee2() {
        var contract;
        return _regenerator["default"].wrap(function _callee2$(_context2) {
          while (1) {
            switch (_context2.prev = _context2.next) {
              case 0:
                _context2.next = 2;
                return this._getManagerContract();

              case 2:
                contract = _context2.sent;
                return _context2.abrupt("return", contract);

              case 4:
              case "end":
                return _context2.stop();
            }
          }
        }, _callee2, this);
      }));

      function getContract() {
        return _getContract.apply(this, arguments);
      }

      return getContract;
    }()
  }, {
    key: "getChild",
    value: function () {
      var _getChild = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee3(stn721Parent, tokenId) {
        var contract;
        return _regenerator["default"].wrap(function _callee3$(_context3) {
          while (1) {
            switch (_context3.prev = _context3.next) {
              case 0:
                _context3.next = 2;
                return this._getManagerContract();

              case 2:
                contract = _context3.sent;
                return _context3.abrupt("return", contract.getChild(stn721Parent, tokenId));

              case 4:
              case "end":
                return _context3.stop();
            }
          }
        }, _callee3, this);
      }));

      function getChild(_x, _x2) {
        return _getChild.apply(this, arguments);
      }

      return getChild;
    }()
  }, {
    key: "getAll",
    value: function () {
      var _getAll = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee4() {
        var contract;
        return _regenerator["default"].wrap(function _callee4$(_context4) {
          while (1) {
            switch (_context4.prev = _context4.next) {
              case 0:
                _context4.next = 2;
                return this._getManagerContract();

              case 2:
                contract = _context4.sent;
                return _context4.abrupt("return", contract.getAll());

              case 4:
              case "end":
                return _context4.stop();
            }
          }
        }, _callee4, this);
      }));

      function getAll() {
        return _getAll.apply(this, arguments);
      }

      return getAll;
    }()
    /** Setters */

  }, {
    key: "addController",
    value: function () {
      var _addController = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee5(newController) {
        var contract;
        return _regenerator["default"].wrap(function _callee5$(_context5) {
          while (1) {
            switch (_context5.prev = _context5.next) {
              case 0:
                _context5.next = 2;
                return this._getManagerContract();

              case 2:
                contract = _context5.sent;
                return _context5.abrupt("return", contract.functions.addController(newController));

              case 4:
              case "end":
                return _context5.stop();
            }
          }
        }, _callee5, this);
      }));

      function addController(_x3) {
        return _addController.apply(this, arguments);
      }

      return addController;
    }()
  }, {
    key: "removeController",
    value: function () {
      var _removeController = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee6(controller) {
        var contract;
        return _regenerator["default"].wrap(function _callee6$(_context6) {
          while (1) {
            switch (_context6.prev = _context6.next) {
              case 0:
                _context6.next = 2;
                return this._getManagerContract();

              case 2:
                contract = _context6.sent;
                return _context6.abrupt("return", contract.functions.removeController(controller));

              case 4:
              case "end":
                return _context6.stop();
            }
          }
        }, _callee6, this);
      }));

      function removeController(_x4) {
        return _removeController.apply(this, arguments);
      }

      return removeController;
    }()
  }, {
    key: "createParentSTN721",
    value: function () {
      var _createParentSTN = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee7(name, symbol) {
        var contract;
        return _regenerator["default"].wrap(function _callee7$(_context7) {
          while (1) {
            switch (_context7.prev = _context7.next) {
              case 0:
                _context7.next = 2;
                return this._getManagerContract();

              case 2:
                contract = _context7.sent;
                return _context7.abrupt("return", contract.functions.createParentSTN721(name, symbol));

              case 4:
              case "end":
                return _context7.stop();
            }
          }
        }, _callee7, this);
      }));

      function createParentSTN721(_x5, _x6) {
        return _createParentSTN.apply(this, arguments);
      }

      return createParentSTN721;
    }()
  }, {
    key: "createChildSTN721",
    value: function () {
      var _createChildSTN = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee8(stn721Parent, tokenId, name, symbol) {
        var contract;
        return _regenerator["default"].wrap(function _callee8$(_context8) {
          while (1) {
            switch (_context8.prev = _context8.next) {
              case 0:
                _context8.next = 2;
                return this._getManagerContract();

              case 2:
                contract = _context8.sent;
                return _context8.abrupt("return", contract.functions.createChildSTN721(stn721Parent, tokenId, name, symbol));

              case 4:
              case "end":
                return _context8.stop();
            }
          }
        }, _callee8, this);
      }));

      function createChildSTN721(_x7, _x8, _x9, _x10) {
        return _createChildSTN.apply(this, arguments);
      }

      return createChildSTN721;
    }()
  }, {
    key: "mint",
    value: function () {
      var _mint = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee9(stn721, divisible, tokenIds, receiver) {
        var contract;
        return _regenerator["default"].wrap(function _callee9$(_context9) {
          while (1) {
            switch (_context9.prev = _context9.next) {
              case 0:
                _context9.next = 2;
                return this._getManagerContract();

              case 2:
                contract = _context9.sent;
                return _context9.abrupt("return", contract.functions.mint(stn721, divisible, tokenIds, receiver));

              case 4:
              case "end":
                return _context9.stop();
            }
          }
        }, _callee9, this);
      }));

      function mint(_x11, _x12, _x13, _x14) {
        return _mint.apply(this, arguments);
      }

      return mint;
    }()
  }, {
    key: "mintChild",
    value: function () {
      var _mintChild = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee10(stn721Child, divisible, tokenIds) {
        var contract;
        return _regenerator["default"].wrap(function _callee10$(_context10) {
          while (1) {
            switch (_context10.prev = _context10.next) {
              case 0:
                _context10.next = 2;
                return this._getManagerContract();

              case 2:
                contract = _context10.sent;
                return _context10.abrupt("return", contract.functions.mintChild(stn721Child, divisible, tokenIds));

              case 4:
              case "end":
                return _context10.stop();
            }
          }
        }, _callee10, this);
      }));

      function mintChild(_x15, _x16, _x17) {
        return _mintChild.apply(this, arguments);
      }

      return mintChild;
    }()
  }], [{
    key: "deploySTN721ManagerContract",
    value: function () {
      var _deploySTN721ManagerContract = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee11(signer, produceAddress, controllers) {
        return _regenerator["default"].wrap(function _callee11$(_context11) {
          while (1) {
            switch (_context11.prev = _context11.next) {
              case 0:
                return _context11.abrupt("return", (0, _deploy.deploySTN721Manager)(signer, produceAddress, controllers));

              case 1:
              case "end":
                return _context11.stop();
            }
          }
        }, _callee11);
      }));

      function deploySTN721ManagerContract(_x18, _x19, _x20) {
        return _deploySTN721ManagerContract.apply(this, arguments);
      }

      return deploySTN721ManagerContract;
    }()
  }]);
  return ManagerController;
}();

exports.ManagerController = ManagerController;
