// smart-trade-networks/lib/controllers/rea/permissionsController.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _typeof = require("@babel/runtime/helpers/typeof");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PermissionsController = void 0;

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _ethers = require("ethers");

var _prefabContractFactory = require("../../chain/prefabContractFactory");

var _deploy = require("../../ethereum/deploy/deploy");

var _encodeCall = require("../../ethereum/encodeCall");

var artifact = _interopRequireWildcard(require("../../ethereum/abi/Permissions.json"));

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

var PermissionsController = /*#__PURE__*/function () {
  function PermissionsController(_permissionsAddress, signerOrProvider) {
    (0, _classCallCheck2["default"])(this, PermissionsController);
    (0, _defineProperty2["default"])(this, "_signer", void 0);
    (0, _defineProperty2["default"])(this, "_permissionsAddress", void 0);
    (0, _defineProperty2["default"])(this, "_permissionsContract", void 0);
    this._permissionsAddress = _permissionsAddress;

    if (signerOrProvider instanceof _ethers.Signer || signerOrProvider instanceof _ethers.ethers.providers.JsonRpcSigner) {
      this._signer = signerOrProvider;
    } else {
      var web3Wrapper = new _ethers.ethers.providers.Web3Provider(signerOrProvider);
      this._signer = web3Wrapper.getSigner();
    }
  }

  (0, _createClass2["default"])(PermissionsController, [{
    key: "_getPermissionsContract",
    value: function () {
      var _getPermissionsContract2 = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee() {
        var contract;
        return _regenerator["default"].wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                if (!this._permissionsContract) {
                  _context.next = 2;
                  break;
                }

                return _context.abrupt("return", this._permissionsContract);

              case 2:
                _context.prev = 2;
                _context.next = 5;
                return (0, _prefabContractFactory.getPermissionsContract)(this._permissionsAddress, this._signer);

              case 5:
                contract = _context.sent;
                this._permissionsContract = contract;
                return _context.abrupt("return", contract);

              case 10:
                _context.prev = 10;
                _context.t0 = _context["catch"](2);
                throw Error("Failed to get Permissions contract: ".concat(_context.t0));

              case 13:
              case "end":
                return _context.stop();
            }
          }
        }, _callee, this, [[2, 10]]);
      }));

      function _getPermissionsContract() {
        return _getPermissionsContract2.apply(this, arguments);
      }

      return _getPermissionsContract;
    }()
  }, {
    key: "getContract",
    value:
    /**Getters */
    function () {
      var _getContract = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee2() {
        var permissionsContract;
        return _regenerator["default"].wrap(function _callee2$(_context2) {
          while (1) {
            switch (_context2.prev = _context2.next) {
              case 0:
                _context2.next = 2;
                return this._getPermissionsContract();

              case 2:
                permissionsContract = _context2.sent;
                return _context2.abrupt("return", permissionsContract);

              case 4:
              case "end":
                return _context2.stop();
            }
          }
        }, _callee2, this);
      }));

      function getContract() {
        return _getContract.apply(this, arguments);
      }

      return getContract;
    }()
  }, {
    key: "getPermission",
    value: function () {
      var _getPermission = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee3(action, companyAddress, nftAddress, signatorie) {
        var permissionsContract;
        return _regenerator["default"].wrap(function _callee3$(_context3) {
          while (1) {
            switch (_context3.prev = _context3.next) {
              case 0:
                _context3.next = 2;
                return this._getPermissionsContract();

              case 2:
                permissionsContract = _context3.sent;
                return _context3.abrupt("return", permissionsContract.getPermission(action, companyAddress, nftAddress, signatorie));

              case 4:
              case "end":
                return _context3.stop();
            }
          }
        }, _callee3, this);
      }));

      function getPermission(_x, _x2, _x3, _x4) {
        return _getPermission.apply(this, arguments);
      }

      return getPermission;
    }()
  }, {
    key: "isSignerApproved",
    value: function () {
      var _isSignerApproved = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee4(companyAddress, signatorie) {
        var permissionsContract;
        return _regenerator["default"].wrap(function _callee4$(_context4) {
          while (1) {
            switch (_context4.prev = _context4.next) {
              case 0:
                _context4.next = 2;
                return this._getPermissionsContract();

              case 2:
                permissionsContract = _context4.sent;
                return _context4.abrupt("return", permissionsContract.isSignerApproved(companyAddress, signatorie));

              case 4:
              case "end":
                return _context4.stop();
            }
          }
        }, _callee4, this);
      }));

      function isSignerApproved(_x5, _x6) {
        return _isSignerApproved.apply(this, arguments);
      }

      return isSignerApproved;
    }()
  }, {
    key: "getCompaniesContract",
    value: function () {
      var _getCompaniesContract = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee5() {
        var permissionsContract;
        return _regenerator["default"].wrap(function _callee5$(_context5) {
          while (1) {
            switch (_context5.prev = _context5.next) {
              case 0:
                _context5.next = 2;
                return this._getPermissionsContract();

              case 2:
                permissionsContract = _context5.sent;
                return _context5.abrupt("return", permissionsContract.getCompaniesContract());

              case 4:
              case "end":
                return _context5.stop();
            }
          }
        }, _callee5, this);
      }));

      function getCompaniesContract() {
        return _getCompaniesContract.apply(this, arguments);
      }

      return getCompaniesContract;
    }()
    /**Setters */

  }, {
    key: "setCompaniesContract",
    value: function () {
      var _setCompaniesContract = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee6(newAddress) {
        var permissionsContract;
        return _regenerator["default"].wrap(function _callee6$(_context6) {
          while (1) {
            switch (_context6.prev = _context6.next) {
              case 0:
                _context6.next = 2;
                return this._getPermissionsContract();

              case 2:
                permissionsContract = _context6.sent;
                return _context6.abrupt("return", permissionsContract.functions.setCompaniesContract(newAddress));

              case 4:
              case "end":
                return _context6.stop();
            }
          }
        }, _callee6, this);
      }));

      function setCompaniesContract(_x7) {
        return _setCompaniesContract.apply(this, arguments);
      }

      return setCompaniesContract;
    }()
  }, {
    key: "setPermission",
    value: function () {
      var _setPermission = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee7(action, companyAddress, nftAddress, allow) {
        var permissionsContract;
        return _regenerator["default"].wrap(function _callee7$(_context7) {
          while (1) {
            switch (_context7.prev = _context7.next) {
              case 0:
                _context7.next = 2;
                return this._getPermissionsContract();

              case 2:
                permissionsContract = _context7.sent;
                return _context7.abrupt("return", permissionsContract.functions.setPermission(action, companyAddress, nftAddress, allow));

              case 4:
              case "end":
                return _context7.stop();
            }
          }
        }, _callee7, this);
      }));

      function setPermission(_x8, _x9, _x10, _x11) {
        return _setPermission.apply(this, arguments);
      }

      return setPermission;
    }()
  }, {
    key: "setPermissions",
    value: function () {
      var _setPermissions = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee8(actions, companyAddress, nftAddress, allow) {
        var permissionsContract;
        return _regenerator["default"].wrap(function _callee8$(_context8) {
          while (1) {
            switch (_context8.prev = _context8.next) {
              case 0:
                _context8.next = 2;
                return this._getPermissionsContract();

              case 2:
                permissionsContract = _context8.sent;
                return _context8.abrupt("return", permissionsContract.functions.setPermissions(actions, companyAddress, nftAddress, allow));

              case 4:
              case "end":
                return _context8.stop();
            }
          }
        }, _callee8, this);
      }));

      function setPermissions(_x12, _x13, _x14, _x15) {
        return _setPermissions.apply(this, arguments);
      }

      return setPermissions;
    }()
  }, {
    key: "setBoolPermissions",
    value: function () {
      var _setBoolPermissions = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee9(permissions, companyAddress, nftAddress) {
        var permissionsContract;
        return _regenerator["default"].wrap(function _callee9$(_context9) {
          while (1) {
            switch (_context9.prev = _context9.next) {
              case 0:
                _context9.next = 2;
                return this._getPermissionsContract();

              case 2:
                permissionsContract = _context9.sent;
                return _context9.abrupt("return", permissionsContract.functions.setBoolPermissions(permissions, companyAddress, nftAddress));

              case 4:
              case "end":
                return _context9.stop();
            }
          }
        }, _callee9, this);
      }));

      function setBoolPermissions(_x16, _x17, _x18) {
        return _setBoolPermissions.apply(this, arguments);
      }

      return setBoolPermissions;
    }()
    /** Helpers */

  }], [{
    key: "deployPermissionsContract",
    value: function () {
      var _deployPermissionsContract = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee10(signer, owner, companiesAddress) {
        return _regenerator["default"].wrap(function _callee10$(_context10) {
          while (1) {
            switch (_context10.prev = _context10.next) {
              case 0:
                return _context10.abrupt("return", (0, _deploy.deployPermissions)(signer, owner, companiesAddress));

              case 1:
              case "end":
                return _context10.stop();
            }
          }
        }, _callee10);
      }));

      function deployPermissionsContract(_x19, _x20, _x21) {
        return _deployPermissionsContract.apply(this, arguments);
      }

      return deployPermissionsContract;
    }()
  }, {
    key: "encodeCall",
    value: function encodeCall(functionName, args) {
      return (0, _encodeCall.encode)(artifact, functionName, args);
    }
  }, {
    key: "decodeCall",
    value: function decodeCall(data) {
      return (0, _encodeCall.decode)(artifact, data);
    }
  }]);
  return PermissionsController;
}();

exports.PermissionsController = PermissionsController;
