// smart-trade-networks/lib/controllers/rea/produceController.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ProduceController = void 0;

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _ethers = require("ethers");

var _prefabContractFactory = require("../../chain/prefabContractFactory");

var _deploy = require("../../ethereum/deploy/deploy");

var ProduceController = /*#__PURE__*/function () {
  function ProduceController(_produceAddress, signerOrProvider) {
    (0, _classCallCheck2["default"])(this, ProduceController);
    (0, _defineProperty2["default"])(this, "_signer", void 0);
    (0, _defineProperty2["default"])(this, "_produceAddress", void 0);
    (0, _defineProperty2["default"])(this, "_produceContract", void 0);
    this._produceAddress = _produceAddress;

    if (signerOrProvider instanceof _ethers.Signer || signerOrProvider instanceof _ethers.ethers.providers.JsonRpcSigner) {
      this._signer = signerOrProvider;
    } else {
      var web3Wrapper = new _ethers.ethers.providers.Web3Provider(signerOrProvider);
      this._signer = web3Wrapper.getSigner();
    }
  }

  (0, _createClass2["default"])(ProduceController, [{
    key: "_getProduceContract",
    value: function () {
      var _getProduceContract2 = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee() {
        var contract;
        return _regenerator["default"].wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                if (!this._produceContract) {
                  _context.next = 2;
                  break;
                }

                return _context.abrupt("return", this._produceContract);

              case 2:
                _context.prev = 2;
                _context.next = 5;
                return (0, _prefabContractFactory.getProduceContract)(this._produceAddress, this._signer);

              case 5:
                contract = _context.sent;
                this._produceContract = contract;
                return _context.abrupt("return", contract);

              case 10:
                _context.prev = 10;
                _context.t0 = _context["catch"](2);
                throw Error("Failed to get Produce contract: ".concat(_context.t0));

              case 13:
              case "end":
                return _context.stop();
            }
          }
        }, _callee, this, [[2, 10]]);
      }));

      function _getProduceContract() {
        return _getProduceContract2.apply(this, arguments);
      }

      return _getProduceContract;
    }()
  }, {
    key: "getContract",
    value:
    /**Getters */
    function () {
      var _getContract = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee2() {
        var produceContract;
        return _regenerator["default"].wrap(function _callee2$(_context2) {
          while (1) {
            switch (_context2.prev = _context2.next) {
              case 0:
                _context2.next = 2;
                return this._getProduceContract();

              case 2:
                produceContract = _context2.sent;
                return _context2.abrupt("return", produceContract);

              case 4:
              case "end":
                return _context2.stop();
            }
          }
        }, _callee2, this);
      }));

      function getContract() {
        return _getContract.apply(this, arguments);
      }

      return getContract;
    }()
  }, {
    key: "getEntriesByCompany",
    value: function () {
      var _getEntriesByCompany = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee3(companyAddress) {
        var produceContract;
        return _regenerator["default"].wrap(function _callee3$(_context3) {
          while (1) {
            switch (_context3.prev = _context3.next) {
              case 0:
                _context3.next = 2;
                return this._getProduceContract();

              case 2:
                produceContract = _context3.sent;
                return _context3.abrupt("return", produceContract.getEntriesByCompany(companyAddress));

              case 4:
              case "end":
                return _context3.stop();
            }
          }
        }, _callee3, this);
      }));

      function getEntriesByCompany(_x) {
        return _getEntriesByCompany.apply(this, arguments);
      }

      return getEntriesByCompany;
    }()
  }, {
    key: "getEntriesByUser",
    value: function () {
      var _getEntriesByUser = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee4(userAddress) {
        var produceContract;
        return _regenerator["default"].wrap(function _callee4$(_context4) {
          while (1) {
            switch (_context4.prev = _context4.next) {
              case 0:
                _context4.next = 2;
                return this._getProduceContract();

              case 2:
                produceContract = _context4.sent;
                return _context4.abrupt("return", produceContract.getEntriesByUser(userAddress));

              case 4:
              case "end":
                return _context4.stop();
            }
          }
        }, _callee4, this);
      }));

      function getEntriesByUser(_x2) {
        return _getEntriesByUser.apply(this, arguments);
      }

      return getEntriesByUser;
    }()
  }, {
    key: "getEntryData",
    value: function () {
      var _getEntryData = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee5(entryNumber) {
        var produceContract;
        return _regenerator["default"].wrap(function _callee5$(_context5) {
          while (1) {
            switch (_context5.prev = _context5.next) {
              case 0:
                _context5.next = 2;
                return this._getProduceContract();

              case 2:
                produceContract = _context5.sent;
                return _context5.abrupt("return", produceContract.getEntryData(entryNumber));

              case 4:
              case "end":
                return _context5.stop();
            }
          }
        }, _callee5, this);
      }));

      function getEntryData(_x3) {
        return _getEntryData.apply(this, arguments);
      }

      return getEntryData;
    }()
  }, {
    key: "getNumOfValidations",
    value: function () {
      var _getNumOfValidations = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee6(entryNumber) {
        var produceContract;
        return _regenerator["default"].wrap(function _callee6$(_context6) {
          while (1) {
            switch (_context6.prev = _context6.next) {
              case 0:
                _context6.next = 2;
                return this._getProduceContract();

              case 2:
                produceContract = _context6.sent;
                return _context6.abrupt("return", produceContract.getNumOfValidations(entryNumber));

              case 4:
              case "end":
                return _context6.stop();
            }
          }
        }, _callee6, this);
      }));

      function getNumOfValidations(_x4) {
        return _getNumOfValidations.apply(this, arguments);
      }

      return getNumOfValidations;
    }()
  }, {
    key: "getValidation",
    value: function () {
      var _getValidation = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee7(entryNumber, index) {
        var produceContract;
        return _regenerator["default"].wrap(function _callee7$(_context7) {
          while (1) {
            switch (_context7.prev = _context7.next) {
              case 0:
                _context7.next = 2;
                return this._getProduceContract();

              case 2:
                produceContract = _context7.sent;
                return _context7.abrupt("return", produceContract.getValidation(entryNumber, index));

              case 4:
              case "end":
                return _context7.stop();
            }
          }
        }, _callee7, this);
      }));

      function getValidation(_x5, _x6) {
        return _getValidation.apply(this, arguments);
      }

      return getValidation;
    }()
    /** Setters */

  }, {
    key: "setPermissionsContract",
    value: function () {
      var _setPermissionsContract = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee8(newAddress) {
        var produceContract;
        return _regenerator["default"].wrap(function _callee8$(_context8) {
          while (1) {
            switch (_context8.prev = _context8.next) {
              case 0:
                _context8.next = 2;
                return this._getProduceContract();

              case 2:
                produceContract = _context8.sent;
                return _context8.abrupt("return", produceContract.functions.setPermissionsContract(newAddress));

              case 4:
              case "end":
                return _context8.stop();
            }
          }
        }, _callee8, this);
      }));

      function setPermissionsContract(_x7) {
        return _setPermissionsContract.apply(this, arguments);
      }

      return setPermissionsContract;
    }()
  }, {
    key: "setSTN721ManagerContract",
    value: function () {
      var _setSTN721ManagerContract = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee9(newAddress) {
        var produceContract;
        return _regenerator["default"].wrap(function _callee9$(_context9) {
          while (1) {
            switch (_context9.prev = _context9.next) {
              case 0:
                _context9.next = 2;
                return this._getProduceContract();

              case 2:
                produceContract = _context9.sent;
                return _context9.abrupt("return", produceContract.functions.setSTN721ManagerContract(newAddress));

              case 4:
              case "end":
                return _context9.stop();
            }
          }
        }, _callee9, this);
      }));

      function setSTN721ManagerContract(_x8) {
        return _setSTN721ManagerContract.apply(this, arguments);
      }

      return setSTN721ManagerContract;
    }()
  }, {
    key: "produce",
    value: function () {
      var _produce = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee10(tokenIds, stn721Address, companyAddress, divisible, involvedCompanies, dbId) {
        var produceContract;
        return _regenerator["default"].wrap(function _callee10$(_context10) {
          while (1) {
            switch (_context10.prev = _context10.next) {
              case 0:
                _context10.next = 2;
                return this._getProduceContract();

              case 2:
                produceContract = _context10.sent;
                return _context10.abrupt("return", produceContract.functions.produce(tokenIds, divisible, stn721Address, companyAddress, involvedCompanies, dbId));

              case 4:
              case "end":
                return _context10.stop();
            }
          }
        }, _callee10, this);
      }));

      function produce(_x9, _x10, _x11, _x12, _x13, _x14) {
        return _produce.apply(this, arguments);
      }

      return produce;
    }()
  }, {
    key: "produceChild",
    value: function () {
      var _produceChild = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee11(tokenIds, stn721Address, companyAddress, divisible, involvedCompanies, dbId) {
        var produceContract;
        return _regenerator["default"].wrap(function _callee11$(_context11) {
          while (1) {
            switch (_context11.prev = _context11.next) {
              case 0:
                _context11.next = 2;
                return this._getProduceContract();

              case 2:
                produceContract = _context11.sent;
                return _context11.abrupt("return", produceContract.functions.produceChild(tokenIds, divisible, stn721Address, companyAddress, involvedCompanies, dbId));

              case 4:
              case "end":
                return _context11.stop();
            }
          }
        }, _callee11, this);
      }));

      function produceChild(_x15, _x16, _x17, _x18, _x19, _x20) {
        return _produceChild.apply(this, arguments);
      }

      return produceChild;
    }()
  }, {
    key: "createSTN721Child",
    value: function () {
      var _createSTN721Child = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee12(companyAddress, stn721Parent, tokenId, name, symbol) {
        var produceContract;
        return _regenerator["default"].wrap(function _callee12$(_context12) {
          while (1) {
            switch (_context12.prev = _context12.next) {
              case 0:
                _context12.next = 2;
                return this._getProduceContract();

              case 2:
                produceContract = _context12.sent;
                return _context12.abrupt("return", produceContract.functions.createSTN721Child(companyAddress, stn721Parent, tokenId, name, symbol));

              case 4:
              case "end":
                return _context12.stop();
            }
          }
        }, _callee12, this);
      }));

      function createSTN721Child(_x21, _x22, _x23, _x24, _x25) {
        return _createSTN721Child.apply(this, arguments);
      }

      return createSTN721Child;
    }()
  }, {
    key: "validateEntry",
    value: function () {
      var _validateEntry = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee13(entryNumber, companyAddress) {
        var produceContract;
        return _regenerator["default"].wrap(function _callee13$(_context13) {
          while (1) {
            switch (_context13.prev = _context13.next) {
              case 0:
                _context13.next = 2;
                return this._getProduceContract();

              case 2:
                produceContract = _context13.sent;
                return _context13.abrupt("return", produceContract.functions.validateEntry(entryNumber, companyAddress));

              case 4:
              case "end":
                return _context13.stop();
            }
          }
        }, _callee13, this);
      }));

      function validateEntry(_x26, _x27) {
        return _validateEntry.apply(this, arguments);
      }

      return validateEntry;
    }()
  }], [{
    key: "deployProduceContract",
    value: function () {
      var _deployProduceContract = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee14(signer, owner, permissionsAddress) {
        return _regenerator["default"].wrap(function _callee14$(_context14) {
          while (1) {
            switch (_context14.prev = _context14.next) {
              case 0:
                return _context14.abrupt("return", (0, _deploy.deployProduce)(signer, owner, permissionsAddress));

              case 1:
              case "end":
                return _context14.stop();
            }
          }
        }, _callee14);
      }));

      function deployProduceContract(_x28, _x29, _x30) {
        return _deployProduceContract.apply(this, arguments);
      }

      return deployProduceContract;
    }()
  }]);
  return ProduceController;
}();

exports.ProduceController = ProduceController;
