// smart-trade-networks/lib/chain/prefabContractFactory.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _typeof = require("@babel/runtime/helpers/typeof");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getManagerContract = getManagerContract;
exports.getSTN721Contract = getSTN721Contract;
exports.getERC721Contract = getERC721Contract;
exports.getMultisigContract = getMultisigContract;
exports.getPermissionsContract = getPermissionsContract;
exports.getProduceContract = getProduceContract;
exports.getCompaniesFactoryContract = getCompaniesFactoryContract;
exports.getCompanyContract = getCompanyContract;
exports.getERC20Contract = getERC20Contract;
exports.getSTN20Contract = getSTN20Contract;
exports.getERC20MaticContract = getERC20MaticContract;
exports.getMaticIRootChain = getMaticIRootChain;
exports.getMultisigDeployerContract = getMultisigDeployerContract;
exports.getMultisigENSContract = getMultisigENSContract;
exports.getMaticRootChainManager = getMaticRootChainManager;
exports.getMaticChildChainManager = getMaticChildChainManager;
exports.getERC20MultisigContract = getERC20MultisigContract;
exports.getCreate2DeployerContract = getCreate2DeployerContract;
exports.getFeeCollectorContract = getFeeCollectorContract;
exports.getOwnableContract = getOwnableContract;

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var STN721ManagerCompiled = _interopRequireWildcard(require("../ethereum/abi/STN721Manager.json"));

var STN721Compiled = _interopRequireWildcard(require("../ethereum/abi/STN721.json"));

var ERC721Compiled = _interopRequireWildcard(require("../ethereum/abi/ERC721Ownable.json"));

var ERC20Compiled = _interopRequireWildcard(require("../ethereum/abi/ERC20.json"));

var STN20Compiled = _interopRequireWildcard(require("../ethereum/abi/STN.json"));

var MultisigCompiled = _interopRequireWildcard(require("../ethereum/abi/MultisigWallet.json"));

var PermissionsCompiled = _interopRequireWildcard(require("../ethereum/abi/Permissions.json"));

var ProduceCompiled = _interopRequireWildcard(require("../ethereum/abi/Produce.json"));

var CompaniesFactoryCompiled = _interopRequireWildcard(require("../ethereum/abi/CompaniesFactory.json"));

var CompanyCompiled = _interopRequireWildcard(require("../ethereum/abi/Company.json"));

var ERC20MaticCompiled = _interopRequireWildcard(require("../ethereum/abi/ChildMintableERC20.json"));

var IRootChainCompiled = _interopRequireWildcard(require("../ethereum/abi/IRootChain.json"));

var MultisigDeployerCompiled = _interopRequireWildcard(require("../ethereum/abi/MultisigDeployer.json"));

var MultisigENSCompiled = _interopRequireWildcard(require("../ethereum/abi/MultisigENS.json"));

var RootChainManagerCompiled = _interopRequireWildcard(require("../ethereum/abi/RootChainManagerStorage.json"));

var ChildChainManagerCompiled = _interopRequireWildcard(require("../ethereum/abi/ChildChainManager.json"));

var ERC20MultisigWalletCompiled = _interopRequireWildcard(require("../ethereum/abi/ERC20MultisigWallet.json"));

var Create2DeployerCompiled = _interopRequireWildcard(require("../ethereum/abi/Create2Deployer.json"));

var FeeCollectorCompiled = _interopRequireWildcard(require("../ethereum/abi/FeeCollector.json"));

var OwnableCompiled = _interopRequireWildcard(require("../ethereum/abi/Ownable.json"));

var _contractFactory = require("./contractFactory");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function getManagerContract(_x, _x2) {
  return _getManagerContract.apply(this, arguments);
}

function _getManagerContract() {
  _getManagerContract = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee(address, signerIdxOrAddressOrSigner) {
    return _regenerator["default"].wrap(function _callee$(_context) {
      while (1) {
        switch (_context.prev = _context.next) {
          case 0:
            return _context.abrupt("return", (0, _contractFactory.getContract)(address, STN721ManagerCompiled.abi, signerIdxOrAddressOrSigner));

          case 1:
          case "end":
            return _context.stop();
        }
      }
    }, _callee);
  }));
  return _getManagerContract.apply(this, arguments);
}

function getSTN721Contract(_x3, _x4) {
  return _getSTN721Contract.apply(this, arguments);
}

function _getSTN721Contract() {
  _getSTN721Contract = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee2(address, signerIdxOrAddressOrSigner) {
    return _regenerator["default"].wrap(function _callee2$(_context2) {
      while (1) {
        switch (_context2.prev = _context2.next) {
          case 0:
            return _context2.abrupt("return", (0, _contractFactory.getContract)(address, STN721Compiled.abi, signerIdxOrAddressOrSigner));

          case 1:
          case "end":
            return _context2.stop();
        }
      }
    }, _callee2);
  }));
  return _getSTN721Contract.apply(this, arguments);
}

function getERC721Contract(_x5, _x6) {
  return _getERC721Contract.apply(this, arguments);
}

function _getERC721Contract() {
  _getERC721Contract = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee3(address, signerIdxOrAddressOrSigner) {
    return _regenerator["default"].wrap(function _callee3$(_context3) {
      while (1) {
        switch (_context3.prev = _context3.next) {
          case 0:
            return _context3.abrupt("return", (0, _contractFactory.getContract)(address, ERC721Compiled.abi, signerIdxOrAddressOrSigner));

          case 1:
          case "end":
            return _context3.stop();
        }
      }
    }, _callee3);
  }));
  return _getERC721Contract.apply(this, arguments);
}

function getMultisigContract(_x7, _x8) {
  return _getMultisigContract.apply(this, arguments);
}

function _getMultisigContract() {
  _getMultisigContract = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee4(address, signerIdxOrAddressOrSigner) {
    return _regenerator["default"].wrap(function _callee4$(_context4) {
      while (1) {
        switch (_context4.prev = _context4.next) {
          case 0:
            return _context4.abrupt("return", (0, _contractFactory.getContract)(address, MultisigCompiled.abi, signerIdxOrAddressOrSigner));

          case 1:
          case "end":
            return _context4.stop();
        }
      }
    }, _callee4);
  }));
  return _getMultisigContract.apply(this, arguments);
}

function getPermissionsContract(_x9, _x10) {
  return _getPermissionsContract.apply(this, arguments);
}

function _getPermissionsContract() {
  _getPermissionsContract = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee5(address, signerIdxOrAddressOrSigner) {
    return _regenerator["default"].wrap(function _callee5$(_context5) {
      while (1) {
        switch (_context5.prev = _context5.next) {
          case 0:
            return _context5.abrupt("return", (0, _contractFactory.getContract)(address, PermissionsCompiled.abi, signerIdxOrAddressOrSigner));

          case 1:
          case "end":
            return _context5.stop();
        }
      }
    }, _callee5);
  }));
  return _getPermissionsContract.apply(this, arguments);
}

function getProduceContract(_x11, _x12) {
  return _getProduceContract.apply(this, arguments);
}

function _getProduceContract() {
  _getProduceContract = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee6(address, signerIdxOrAddressOrSigner) {
    return _regenerator["default"].wrap(function _callee6$(_context6) {
      while (1) {
        switch (_context6.prev = _context6.next) {
          case 0:
            return _context6.abrupt("return", (0, _contractFactory.getContract)(address, ProduceCompiled.abi, signerIdxOrAddressOrSigner));

          case 1:
          case "end":
            return _context6.stop();
        }
      }
    }, _callee6);
  }));
  return _getProduceContract.apply(this, arguments);
}

function getCompaniesFactoryContract(_x13, _x14) {
  return _getCompaniesFactoryContract.apply(this, arguments);
}

function _getCompaniesFactoryContract() {
  _getCompaniesFactoryContract = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee7(address, signerIdxOrAddressOrSigner) {
    return _regenerator["default"].wrap(function _callee7$(_context7) {
      while (1) {
        switch (_context7.prev = _context7.next) {
          case 0:
            return _context7.abrupt("return", (0, _contractFactory.getContract)(address, CompaniesFactoryCompiled.abi, signerIdxOrAddressOrSigner));

          case 1:
          case "end":
            return _context7.stop();
        }
      }
    }, _callee7);
  }));
  return _getCompaniesFactoryContract.apply(this, arguments);
}

function getCompanyContract(_x15, _x16) {
  return _getCompanyContract.apply(this, arguments);
}

function _getCompanyContract() {
  _getCompanyContract = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee8(address, signerIdxOrAddressOrSigner) {
    return _regenerator["default"].wrap(function _callee8$(_context8) {
      while (1) {
        switch (_context8.prev = _context8.next) {
          case 0:
            return _context8.abrupt("return", (0, _contractFactory.getContract)(address, CompanyCompiled.abi, signerIdxOrAddressOrSigner));

          case 1:
          case "end":
            return _context8.stop();
        }
      }
    }, _callee8);
  }));
  return _getCompanyContract.apply(this, arguments);
}

function getERC20Contract(_x17, _x18) {
  return _getERC20Contract.apply(this, arguments);
}

function _getERC20Contract() {
  _getERC20Contract = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee9(address, signerIdxOrAddressOrSigner) {
    return _regenerator["default"].wrap(function _callee9$(_context9) {
      while (1) {
        switch (_context9.prev = _context9.next) {
          case 0:
            return _context9.abrupt("return", (0, _contractFactory.getContract)(address, ERC20Compiled.abi, signerIdxOrAddressOrSigner));

          case 1:
          case "end":
            return _context9.stop();
        }
      }
    }, _callee9);
  }));
  return _getERC20Contract.apply(this, arguments);
}

function getSTN20Contract(_x19, _x20) {
  return _getSTN20Contract.apply(this, arguments);
}

function _getSTN20Contract() {
  _getSTN20Contract = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee10(address, signerIdxOrAddressOrSigner) {
    return _regenerator["default"].wrap(function _callee10$(_context10) {
      while (1) {
        switch (_context10.prev = _context10.next) {
          case 0:
            return _context10.abrupt("return", (0, _contractFactory.getContract)(address, STN20Compiled.abi, signerIdxOrAddressOrSigner));

          case 1:
          case "end":
            return _context10.stop();
        }
      }
    }, _callee10);
  }));
  return _getSTN20Contract.apply(this, arguments);
}

function getERC20MaticContract(_x21, _x22) {
  return _getERC20MaticContract.apply(this, arguments);
}

function _getERC20MaticContract() {
  _getERC20MaticContract = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee11(address, signerIdxOrAddressOrSigner) {
    return _regenerator["default"].wrap(function _callee11$(_context11) {
      while (1) {
        switch (_context11.prev = _context11.next) {
          case 0:
            return _context11.abrupt("return", (0, _contractFactory.getContract)(address, ERC20MaticCompiled.abi, signerIdxOrAddressOrSigner));

          case 1:
          case "end":
            return _context11.stop();
        }
      }
    }, _callee11);
  }));
  return _getERC20MaticContract.apply(this, arguments);
}

function getMaticIRootChain(_x23, _x24) {
  return _getMaticIRootChain.apply(this, arguments);
}

function _getMaticIRootChain() {
  _getMaticIRootChain = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee12(address, signerIdxOrAddressOrSigner) {
    return _regenerator["default"].wrap(function _callee12$(_context12) {
      while (1) {
        switch (_context12.prev = _context12.next) {
          case 0:
            return _context12.abrupt("return", (0, _contractFactory.getContract)(address, IRootChainCompiled.abi, signerIdxOrAddressOrSigner));

          case 1:
          case "end":
            return _context12.stop();
        }
      }
    }, _callee12);
  }));
  return _getMaticIRootChain.apply(this, arguments);
}

function getMultisigDeployerContract(_x25, _x26) {
  return _getMultisigDeployerContract.apply(this, arguments);
}

function _getMultisigDeployerContract() {
  _getMultisigDeployerContract = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee13(address, signerIdxOrAddressOrSigner) {
    return _regenerator["default"].wrap(function _callee13$(_context13) {
      while (1) {
        switch (_context13.prev = _context13.next) {
          case 0:
            return _context13.abrupt("return", (0, _contractFactory.getContract)(address, MultisigDeployerCompiled.abi, signerIdxOrAddressOrSigner));

          case 1:
          case "end":
            return _context13.stop();
        }
      }
    }, _callee13);
  }));
  return _getMultisigDeployerContract.apply(this, arguments);
}

function getMultisigENSContract(_x27, _x28) {
  return _getMultisigENSContract.apply(this, arguments);
}

function _getMultisigENSContract() {
  _getMultisigENSContract = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee14(address, signerIdxOrAddressOrSigner) {
    return _regenerator["default"].wrap(function _callee14$(_context14) {
      while (1) {
        switch (_context14.prev = _context14.next) {
          case 0:
            return _context14.abrupt("return", (0, _contractFactory.getContract)(address, MultisigENSCompiled.abi, signerIdxOrAddressOrSigner));

          case 1:
          case "end":
            return _context14.stop();
        }
      }
    }, _callee14);
  }));
  return _getMultisigENSContract.apply(this, arguments);
}

function getMaticRootChainManager(_x29, _x30) {
  return _getMaticRootChainManager.apply(this, arguments);
}

function _getMaticRootChainManager() {
  _getMaticRootChainManager = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee15(address, signerIdxOrAddressOrSigner) {
    return _regenerator["default"].wrap(function _callee15$(_context15) {
      while (1) {
        switch (_context15.prev = _context15.next) {
          case 0:
            return _context15.abrupt("return", (0, _contractFactory.getContract)(address, RootChainManagerCompiled.abi, signerIdxOrAddressOrSigner));

          case 1:
          case "end":
            return _context15.stop();
        }
      }
    }, _callee15);
  }));
  return _getMaticRootChainManager.apply(this, arguments);
}

function getMaticChildChainManager(_x31, _x32) {
  return _getMaticChildChainManager.apply(this, arguments);
}

function _getMaticChildChainManager() {
  _getMaticChildChainManager = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee16(address, signerIdxOrAddressOrSigner) {
    return _regenerator["default"].wrap(function _callee16$(_context16) {
      while (1) {
        switch (_context16.prev = _context16.next) {
          case 0:
            return _context16.abrupt("return", (0, _contractFactory.getContract)(address, ChildChainManagerCompiled.abi, signerIdxOrAddressOrSigner));

          case 1:
          case "end":
            return _context16.stop();
        }
      }
    }, _callee16);
  }));
  return _getMaticChildChainManager.apply(this, arguments);
}

function getERC20MultisigContract(_x33, _x34) {
  return _getERC20MultisigContract.apply(this, arguments);
}

function _getERC20MultisigContract() {
  _getERC20MultisigContract = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee17(address, signerIdxOrAddressOrSigner) {
    return _regenerator["default"].wrap(function _callee17$(_context17) {
      while (1) {
        switch (_context17.prev = _context17.next) {
          case 0:
            return _context17.abrupt("return", (0, _contractFactory.getContract)(address, ERC20MultisigWalletCompiled.abi, signerIdxOrAddressOrSigner));

          case 1:
          case "end":
            return _context17.stop();
        }
      }
    }, _callee17);
  }));
  return _getERC20MultisigContract.apply(this, arguments);
}

function getCreate2DeployerContract(_x35, _x36) {
  return _getCreate2DeployerContract.apply(this, arguments);
}

function _getCreate2DeployerContract() {
  _getCreate2DeployerContract = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee18(address, signerIdxOrAddressOrSigner) {
    return _regenerator["default"].wrap(function _callee18$(_context18) {
      while (1) {
        switch (_context18.prev = _context18.next) {
          case 0:
            return _context18.abrupt("return", (0, _contractFactory.getContract)(address, Create2DeployerCompiled.abi, signerIdxOrAddressOrSigner));

          case 1:
          case "end":
            return _context18.stop();
        }
      }
    }, _callee18);
  }));
  return _getCreate2DeployerContract.apply(this, arguments);
}

function getFeeCollectorContract(_x37, _x38) {
  return _getFeeCollectorContract.apply(this, arguments);
}

function _getFeeCollectorContract() {
  _getFeeCollectorContract = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee19(address, signerIdxOrAddressOrSigner) {
    return _regenerator["default"].wrap(function _callee19$(_context19) {
      while (1) {
        switch (_context19.prev = _context19.next) {
          case 0:
            return _context19.abrupt("return", (0, _contractFactory.getContract)(address, FeeCollectorCompiled.abi, signerIdxOrAddressOrSigner));

          case 1:
          case "end":
            return _context19.stop();
        }
      }
    }, _callee19);
  }));
  return _getFeeCollectorContract.apply(this, arguments);
}

function getOwnableContract(_x39, _x40) {
  return _getOwnableContract.apply(this, arguments);
}

function _getOwnableContract() {
  _getOwnableContract = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee20(address, signerIdxOrAddressOrSigner) {
    return _regenerator["default"].wrap(function _callee20$(_context20) {
      while (1) {
        switch (_context20.prev = _context20.next) {
          case 0:
            return _context20.abrupt("return", (0, _contractFactory.getContract)(address, OwnableCompiled.abi, signerIdxOrAddressOrSigner));

          case 1:
          case "end":
            return _context20.stop();
        }
      }
    }, _callee20);
  }));
  return _getOwnableContract.apply(this, arguments);
}
