// smart-trade-networks/lib/ethereum/deploy/deploy.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _typeof = require("@babel/runtime/helpers/typeof");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.deploySTN721 = deploySTN721;
exports.deploySTN721Manager = deploySTN721Manager;
exports.deployMultisig = deployMultisig;
exports.deployPermissions = deployPermissions;
exports.deployProduce = deployProduce;
exports.deployCompaniesfactory = deployCompaniesfactory;
exports.initializeMultisigDeployer = initializeMultisigDeployer;

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _contractFactory = require("../../chain/contractFactory");

var artifactSTN721 = _interopRequireWildcard(require("../abi/STN721.json"));

var artifactSTN721Manager = _interopRequireWildcard(require("../abi/STN721Manager.json"));

var artifactMultisig = _interopRequireWildcard(require("../abi/MultisigWallet.json"));

var artifactPermissions = _interopRequireWildcard(require("../abi/Permissions.json"));

var artifactCompaniesFactory = _interopRequireWildcard(require("../abi/CompaniesFactory.json"));

var artifactMultisigDeployer = _interopRequireWildcard(require("../abi/MultisigDeployer.json"));

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function deploySTN721(_x, _x2, _x3, _x4, _x5) {
  return _deploySTN.apply(this, arguments);
}

function _deploySTN() {
  _deploySTN = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee(signer, parentAddress, parentId, name, symbol) {
    var contract;
    return _regenerator["default"].wrap(function _callee$(_context) {
      while (1) {
        switch (_context.prev = _context.next) {
          case 0:
            _context.next = 2;
            return (0, _contractFactory.deployContract)(signer, artifactSTN721, parentAddress, parentId, name, symbol);

          case 2:
            contract = _context.sent;
            return _context.abrupt("return", contract);

          case 4:
          case "end":
            return _context.stop();
        }
      }
    }, _callee);
  }));
  return _deploySTN.apply(this, arguments);
}

function deploySTN721Manager(_x6, _x7, _x8) {
  return _deploySTN721Manager.apply(this, arguments);
}

function _deploySTN721Manager() {
  _deploySTN721Manager = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee2(signer, produceAddress, controllers) {
    var contract;
    return _regenerator["default"].wrap(function _callee2$(_context2) {
      while (1) {
        switch (_context2.prev = _context2.next) {
          case 0:
            _context2.next = 2;
            return (0, _contractFactory.deployContract)(signer, artifactSTN721Manager, produceAddress, controllers);

          case 2:
            contract = _context2.sent;
            return _context2.abrupt("return", contract);

          case 4:
          case "end":
            return _context2.stop();
        }
      }
    }, _callee2);
  }));
  return _deploySTN721Manager.apply(this, arguments);
}

function deployMultisig(_x9, _x10, _x11) {
  return _deployMultisig.apply(this, arguments);
}
/** Permissions */


function _deployMultisig() {
  _deployMultisig = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee3(signer, owners, requirement) {
    var contract;
    return _regenerator["default"].wrap(function _callee3$(_context3) {
      while (1) {
        switch (_context3.prev = _context3.next) {
          case 0:
            _context3.next = 2;
            return (0, _contractFactory.deployContract)(signer, artifactMultisig, owners, requirement);

          case 2:
            contract = _context3.sent;
            return _context3.abrupt("return", contract);

          case 4:
          case "end":
            return _context3.stop();
        }
      }
    }, _callee3);
  }));
  return _deployMultisig.apply(this, arguments);
}

function deployPermissions(_x12, _x13, _x14) {
  return _deployPermissions.apply(this, arguments);
}
/** Produce */


function _deployPermissions() {
  _deployPermissions = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee4(signer, owner, companiesAddress) {
    var contract;
    return _regenerator["default"].wrap(function _callee4$(_context4) {
      while (1) {
        switch (_context4.prev = _context4.next) {
          case 0:
            _context4.next = 2;
            return (0, _contractFactory.deployContract)(signer, artifactPermissions, owner, companiesAddress);

          case 2:
            contract = _context4.sent;
            return _context4.abrupt("return", contract);

          case 4:
          case "end":
            return _context4.stop();
        }
      }
    }, _callee4);
  }));
  return _deployPermissions.apply(this, arguments);
}

function deployProduce(_x15, _x16, _x17) {
  return _deployProduce.apply(this, arguments);
}
/** CompaniesFactory */


function _deployProduce() {
  _deployProduce = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee5(signer, owner, permissionsAddress) {
    var contract;
    return _regenerator["default"].wrap(function _callee5$(_context5) {
      while (1) {
        switch (_context5.prev = _context5.next) {
          case 0:
            _context5.next = 2;
            return (0, _contractFactory.deployContract)(signer, artifactPermissions, owner, permissionsAddress);

          case 2:
            contract = _context5.sent;
            return _context5.abrupt("return", contract);

          case 4:
          case "end":
            return _context5.stop();
        }
      }
    }, _callee5);
  }));
  return _deployProduce.apply(this, arguments);
}

function deployCompaniesfactory(_x18, _x19) {
  return _deployCompaniesfactory.apply(this, arguments);
}
/** MultisigDeployer */


function _deployCompaniesfactory() {
  _deployCompaniesfactory = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee6(signer, owner) {
    var contract;
    return _regenerator["default"].wrap(function _callee6$(_context6) {
      while (1) {
        switch (_context6.prev = _context6.next) {
          case 0:
            _context6.next = 2;
            return (0, _contractFactory.deployContract)(signer, artifactCompaniesFactory, owner);

          case 2:
            contract = _context6.sent;
            return _context6.abrupt("return", contract);

          case 4:
          case "end":
            return _context6.stop();
        }
      }
    }, _callee6);
  }));
  return _deployCompaniesfactory.apply(this, arguments);
}

function initializeMultisigDeployer(_x20, _x21) {
  return _initializeMultisigDeployer.apply(this, arguments);
}

function _initializeMultisigDeployer() {
  _initializeMultisigDeployer = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee7(signer, owner) {
    var _yield$initializeCont, _yield$initializeCont2, contract, proxyAddress;

    return _regenerator["default"].wrap(function _callee7$(_context7) {
      while (1) {
        switch (_context7.prev = _context7.next) {
          case 0:
            _context7.next = 2;
            return (0, _contractFactory.initializeContract)(signer, artifactMultisigDeployer, "initialize", [owner]);

          case 2:
            _yield$initializeCont = _context7.sent;
            _yield$initializeCont2 = (0, _slicedToArray2["default"])(_yield$initializeCont, 2);
            contract = _yield$initializeCont2[0];
            proxyAddress = _yield$initializeCont2[1];
            return _context7.abrupt("return", [contract, proxyAddress]);

          case 7:
          case "end":
            return _context7.stop();
        }
      }
    }, _callee7);
  }));
  return _initializeMultisigDeployer.apply(this, arguments);
}
