// smart-trade-networks/lib/types/factories/ChildChainManager__factory.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ChildChainManager__factory = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _ethers = require("ethers");

/* Autogenerated file. Do not edit manually. */

/* tslint:disable */

/* eslint-disable */
var ChildChainManager__factory = /*#__PURE__*/function () {
  function ChildChainManager__factory() {
    (0, _classCallCheck2["default"])(this, ChildChainManager__factory);
  }

  (0, _createClass2["default"])(ChildChainManager__factory, null, [{
    key: "connect",
    value: function connect(address, signerOrProvider) {
      return new _ethers.Contract(address, _abi, signerOrProvider);
    }
  }]);
  return ChildChainManager__factory;
}();

exports.ChildChainManager__factory = ChildChainManager__factory;
var _abi = [{
  inputs: [{
    internalType: "address",
    name: "",
    type: "address"
  }],
  name: "childToRootToken",
  outputs: [{
    internalType: "address",
    name: "",
    type: "address"
  }],
  stateMutability: "view",
  type: "function"
}, {
  inputs: [{
    internalType: "address",
    name: "",
    type: "address"
  }],
  name: "rootToChildToken",
  outputs: [{
    internalType: "address",
    name: "",
    type: "address"
  }],
  stateMutability: "view",
  type: "function"
}];
