// smart-trade-networks/lib/types/factories/IChildToken__factory.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.IChildToken__factory = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _ethers = require("ethers");

/* Autogenerated file. Do not edit manually. */

/* tslint:disable */

/* eslint-disable */
var IChildToken__factory = /*#__PURE__*/function () {
  function IChildToken__factory() {
    (0, _classCallCheck2["default"])(this, IChildToken__factory);
  }

  (0, _createClass2["default"])(IChildToken__factory, null, [{
    key: "connect",
    value: function connect(address, signerOrProvider) {
      return new _ethers.Contract(address, _abi, signerOrProvider);
    }
  }]);
  return IChildToken__factory;
}();

exports.IChildToken__factory = IChildToken__factory;
var _abi = [{
  inputs: [{
    internalType: "address",
    name: "user",
    type: "address"
  }, {
    internalType: "bytes",
    name: "depositData",
    type: "bytes"
  }],
  name: "deposit",
  outputs: [],
  stateMutability: "nonpayable",
  type: "function"
}];
