// smart-trade-networks/lib/types/factories/ICompanies__factory.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ICompanies__factory = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _ethers = require("ethers");

/* Autogenerated file. Do not edit manually. */

/* tslint:disable */

/* eslint-disable */
var ICompanies__factory = /*#__PURE__*/function () {
  function ICompanies__factory() {
    (0, _classCallCheck2["default"])(this, ICompanies__factory);
  }

  (0, _createClass2["default"])(ICompanies__factory, null, [{
    key: "connect",
    value: function connect(address, signerOrProvider) {
      return new _ethers.Contract(address, _abi, signerOrProvider);
    }
  }]);
  return ICompanies__factory;
}();

exports.ICompanies__factory = ICompanies__factory;
var _abi = [{
  anonymous: false,
  inputs: [{
    indexed: false,
    internalType: "bytes32",
    name: "lastIPFS",
    type: "bytes32"
  }, {
    indexed: false,
    internalType: "bytes32",
    name: "newIPFS",
    type: "bytes32"
  }],
  name: "ChangeCompanyIPFS",
  type: "event"
}, {
  anonymous: false,
  inputs: [{
    indexed: false,
    internalType: "string",
    name: "lastName",
    type: "string"
  }, {
    indexed: false,
    internalType: "string",
    name: "newName",
    type: "string"
  }],
  name: "ChangeCompanyName",
  type: "event"
}, {
  anonymous: false,
  inputs: [{
    indexed: true,
    internalType: "address",
    name: "company",
    type: "address"
  }, {
    indexed: false,
    internalType: "string",
    name: "name",
    type: "string"
  }],
  name: "NewCompany",
  type: "event"
}, {
  anonymous: false,
  inputs: [{
    indexed: true,
    internalType: "address",
    name: "company",
    type: "address"
  }, {
    indexed: false,
    internalType: "address",
    name: "signatorie",
    type: "address"
  }],
  name: "NewSignatorie",
  type: "event"
}, {
  anonymous: false,
  inputs: [{
    indexed: true,
    internalType: "address",
    name: "company",
    type: "address"
  }],
  name: "RemoveCompany",
  type: "event"
}, {
  anonymous: false,
  inputs: [{
    indexed: true,
    internalType: "address",
    name: "company",
    type: "address"
  }, {
    indexed: false,
    internalType: "address",
    name: "signatorie",
    type: "address"
  }],
  name: "RemoveSignatorie",
  type: "event"
}, {
  inputs: [{
    internalType: "address",
    name: "_company",
    type: "address"
  }, {
    internalType: "string",
    name: "_name",
    type: "string"
  }, {
    internalType: "bytes32",
    name: "_ipfsLink",
    type: "bytes32"
  }],
  name: "addCompany",
  outputs: [],
  stateMutability: "nonpayable",
  type: "function"
}, {
  inputs: [{
    internalType: "address",
    name: "_company",
    type: "address"
  }],
  name: "removeCompany",
  outputs: [],
  stateMutability: "nonpayable",
  type: "function"
}, {
  inputs: [{
    internalType: "address",
    name: "_signatorie",
    type: "address"
  }],
  name: "addSignatorie",
  outputs: [],
  stateMutability: "nonpayable",
  type: "function"
}, {
  inputs: [{
    internalType: "address",
    name: "_signatorie",
    type: "address"
  }],
  name: "removeSignatorie",
  outputs: [],
  stateMutability: "nonpayable",
  type: "function"
}, {
  inputs: [{
    internalType: "string",
    name: "_name",
    type: "string"
  }, {
    internalType: "bytes32",
    name: "_ipfsLink",
    type: "bytes32"
  }],
  name: "setCompanyData",
  outputs: [],
  stateMutability: "nonpayable",
  type: "function"
}, {
  inputs: [],
  name: "getAllCompanies",
  outputs: [{
    internalType: "address[]",
    name: "",
    type: "address[]"
  }],
  stateMutability: "view",
  type: "function"
}, {
  inputs: [{
    internalType: "address",
    name: "_company",
    type: "address"
  }],
  name: "isAddressCompany",
  outputs: [{
    internalType: "bool",
    name: "",
    type: "bool"
  }],
  stateMutability: "view",
  type: "function"
}, {
  inputs: [{
    internalType: "address",
    name: "_signatorie",
    type: "address"
  }],
  name: "getCompanyBySignatorie",
  outputs: [{
    internalType: "address",
    name: "",
    type: "address"
  }],
  stateMutability: "view",
  type: "function"
}, {
  inputs: [{
    internalType: "address",
    name: "_company",
    type: "address"
  }],
  name: "getCompanyByAddress",
  outputs: [{
    components: [{
      internalType: "uint256",
      name: "createdAt",
      type: "uint256"
    }, {
      internalType: "address[]",
      name: "signatories",
      type: "address[]"
    }, {
      internalType: "string",
      name: "name",
      type: "string"
    }, {
      internalType: "bytes32",
      name: "ipfsLink",
      type: "bytes32"
    }],
    internalType: "struct ICompanies.CompanyData",
    name: "",
    type: "tuple"
  }, {
    internalType: "bool",
    name: "",
    type: "bool"
  }],
  stateMutability: "view",
  type: "function"
}, {
  inputs: [{
    internalType: "string",
    name: "_stringName",
    type: "string"
  }],
  name: "getCompanyByName",
  outputs: [{
    components: [{
      internalType: "uint256",
      name: "createdAt",
      type: "uint256"
    }, {
      internalType: "address[]",
      name: "signatories",
      type: "address[]"
    }, {
      internalType: "string",
      name: "name",
      type: "string"
    }, {
      internalType: "bytes32",
      name: "ipfsLink",
      type: "bytes32"
    }],
    internalType: "struct ICompanies.CompanyData",
    name: "",
    type: "tuple"
  }, {
    internalType: "bool",
    name: "",
    type: "bool"
  }],
  stateMutability: "view",
  type: "function"
}, {
  inputs: [{
    internalType: "address",
    name: "_signatorie",
    type: "address"
  }],
  name: "isSignatorieAuthorized",
  outputs: [{
    internalType: "bool",
    name: "",
    type: "bool"
  }],
  stateMutability: "view",
  type: "function"
}];
