// smart-trade-networks/lib/types/factories/ICompaniesFactory__factory.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ICompaniesFactory__factory = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _ethers = require("ethers");

/* Autogenerated file. Do not edit manually. */

/* tslint:disable */

/* eslint-disable */
var ICompaniesFactory__factory = /*#__PURE__*/function () {
  function ICompaniesFactory__factory() {
    (0, _classCallCheck2["default"])(this, ICompaniesFactory__factory);
  }

  (0, _createClass2["default"])(ICompaniesFactory__factory, null, [{
    key: "connect",
    value: function connect(address, signerOrProvider) {
      return new _ethers.Contract(address, _abi, signerOrProvider);
    }
  }]);
  return ICompaniesFactory__factory;
}();

exports.ICompaniesFactory__factory = ICompaniesFactory__factory;
var _abi = [{
  anonymous: false,
  inputs: [{
    indexed: true,
    internalType: "address",
    name: "company",
    type: "address"
  }],
  name: "ApproveCompany",
  type: "event"
}, {
  anonymous: false,
  inputs: [{
    indexed: true,
    internalType: "address",
    name: "companyAddress",
    type: "address"
  }, {
    indexed: true,
    internalType: "address",
    name: "companyOwner",
    type: "address"
  }],
  name: "NewCompany",
  type: "event"
}, {
  anonymous: false,
  inputs: [{
    indexed: true,
    internalType: "address",
    name: "company",
    type: "address"
  }, {
    indexed: true,
    internalType: "address",
    name: "newSignatorie",
    type: "address"
  }],
  name: "NewSignatorie",
  type: "event"
}, {
  anonymous: false,
  inputs: [{
    indexed: true,
    internalType: "address",
    name: "company",
    type: "address"
  }],
  name: "RemoveCompany",
  type: "event"
}, {
  anonymous: false,
  inputs: [{
    indexed: true,
    internalType: "address",
    name: "company",
    type: "address"
  }, {
    indexed: true,
    internalType: "address",
    name: "signatorie",
    type: "address"
  }],
  name: "RemoveSignatorie",
  type: "event"
}, {
  inputs: [{
    internalType: "address",
    name: "_owner",
    type: "address"
  }, {
    internalType: "string",
    name: "_name",
    type: "string"
  }, {
    internalType: "bytes32",
    name: "_ipfsLink",
    type: "bytes32"
  }],
  name: "createCompany",
  outputs: [],
  stateMutability: "nonpayable",
  type: "function"
}, {
  inputs: [{
    internalType: "address",
    name: "_company",
    type: "address"
  }],
  name: "removeCompany",
  outputs: [],
  stateMutability: "nonpayable",
  type: "function"
}, {
  inputs: [{
    internalType: "address",
    name: "_company",
    type: "address"
  }],
  name: "approveCompany",
  outputs: [],
  stateMutability: "nonpayable",
  type: "function"
}, {
  inputs: [{
    internalType: "address",
    name: "_company",
    type: "address"
  }, {
    internalType: "address",
    name: "_newSignatorie",
    type: "address"
  }],
  name: "addCompanySignatorie",
  outputs: [],
  stateMutability: "nonpayable",
  type: "function"
}, {
  inputs: [{
    internalType: "address",
    name: "_company",
    type: "address"
  }, {
    internalType: "address",
    name: "_signatorie",
    type: "address"
  }],
  name: "removeCompanySignatorie",
  outputs: [],
  stateMutability: "nonpayable",
  type: "function"
}, {
  inputs: [],
  name: "getCompanies",
  outputs: [{
    internalType: "address[]",
    name: "",
    type: "address[]"
  }],
  stateMutability: "view",
  type: "function"
}, {
  inputs: [{
    internalType: "address",
    name: "_company",
    type: "address"
  }, {
    internalType: "address",
    name: "_owner",
    type: "address"
  }],
  name: "isCompanyOwner",
  outputs: [{
    internalType: "bool",
    name: "",
    type: "bool"
  }],
  stateMutability: "view",
  type: "function"
}, {
  inputs: [{
    internalType: "address",
    name: "_signatorie",
    type: "address"
  }],
  name: "getCompaniesBySignatorie",
  outputs: [{
    internalType: "address[]",
    name: "",
    type: "address[]"
  }],
  stateMutability: "view",
  type: "function"
}, {
  inputs: [{
    internalType: "address",
    name: "_company",
    type: "address"
  }],
  name: "getSignersByCompany",
  outputs: [{
    internalType: "address[]",
    name: "",
    type: "address[]"
  }],
  stateMutability: "view",
  type: "function"
}, {
  inputs: [{
    internalType: "address",
    name: "_company",
    type: "address"
  }],
  name: "isCompanyApproved",
  outputs: [{
    internalType: "bool",
    name: "",
    type: "bool"
  }],
  stateMutability: "view",
  type: "function"
}, {
  inputs: [{
    internalType: "address",
    name: "_company",
    type: "address"
  }, {
    internalType: "address",
    name: "_signer",
    type: "address"
  }],
  name: "isSignerApproved",
  outputs: [{
    internalType: "bool",
    name: "",
    type: "bool"
  }],
  stateMutability: "view",
  type: "function"
}];
