// smart-trade-networks/lib/types/factories/ICompany__factory.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ICompany__factory = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _ethers = require("ethers");

/* Autogenerated file. Do not edit manually. */

/* tslint:disable */

/* eslint-disable */
var ICompany__factory = /*#__PURE__*/function () {
  function ICompany__factory() {
    (0, _classCallCheck2["default"])(this, ICompany__factory);
  }

  (0, _createClass2["default"])(ICompany__factory, null, [{
    key: "connect",
    value: function connect(address, signerOrProvider) {
      return new _ethers.Contract(address, _abi, signerOrProvider);
    }
  }]);
  return ICompany__factory;
}();

exports.ICompany__factory = ICompany__factory;
var _abi = [{
  anonymous: false,
  inputs: [{
    indexed: false,
    internalType: "bytes32",
    name: "oldIpfsLink",
    type: "bytes32"
  }, {
    indexed: false,
    internalType: "bytes32",
    name: "newIpfsLink",
    type: "bytes32"
  }],
  name: "ChangeIpfsLink",
  type: "event"
}, {
  anonymous: false,
  inputs: [{
    indexed: false,
    internalType: "string",
    name: "oldName",
    type: "string"
  }, {
    indexed: false,
    internalType: "string",
    name: "newName",
    type: "string"
  }],
  name: "ChangeName",
  type: "event"
}, {
  anonymous: false,
  inputs: [{
    indexed: true,
    internalType: "address",
    name: "newSignatorie",
    type: "address"
  }],
  name: "NewSignatorie",
  type: "event"
}, {
  anonymous: false,
  inputs: [{
    indexed: true,
    internalType: "address",
    name: "signatorie",
    type: "address"
  }],
  name: "RemovedSignatorie",
  type: "event"
}, {
  inputs: [{
    internalType: "string",
    name: "_name",
    type: "string"
  }],
  name: "changeName",
  outputs: [],
  stateMutability: "nonpayable",
  type: "function"
}, {
  inputs: [{
    internalType: "bytes32",
    name: "_ipfsLink",
    type: "bytes32"
  }],
  name: "changeIpfsLink",
  outputs: [],
  stateMutability: "nonpayable",
  type: "function"
}, {
  inputs: [],
  name: "getData",
  outputs: [{
    components: [{
      internalType: "string",
      name: "name",
      type: "string"
    }, {
      internalType: "bytes32",
      name: "ipfsLink",
      type: "bytes32"
    }, {
      internalType: "uint256",
      name: "createdAt",
      type: "uint256"
    }, {
      internalType: "address",
      name: "owner",
      type: "address"
    }],
    internalType: "struct ICompany.CompanyData",
    name: "",
    type: "tuple"
  }],
  stateMutability: "view",
  type: "function"
}];
