// smart-trade-networks/lib/types/factories/IRootChain__factory.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.IRootChain__factory = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _ethers = require("ethers");

/* Autogenerated file. Do not edit manually. */

/* tslint:disable */

/* eslint-disable */
var IRootChain__factory = /*#__PURE__*/function () {
  function IRootChain__factory() {
    (0, _classCallCheck2["default"])(this, IRootChain__factory);
  }

  (0, _createClass2["default"])(IRootChain__factory, null, [{
    key: "connect",
    value: function connect(address, signerOrProvider) {
      return new _ethers.Contract(address, _abi, signerOrProvider);
    }
  }]);
  return IRootChain__factory;
}();

exports.IRootChain__factory = IRootChain__factory;
var _abi = [{
  anonymous: false,
  inputs: [{
    indexed: true,
    internalType: "address",
    name: "proposer",
    type: "address"
  }, {
    indexed: true,
    internalType: "uint256",
    name: "headerBlockId",
    type: "uint256"
  }, {
    indexed: true,
    internalType: "uint256",
    name: "reward",
    type: "uint256"
  }, {
    indexed: false,
    internalType: "uint256",
    name: "start",
    type: "uint256"
  }, {
    indexed: false,
    internalType: "uint256",
    name: "end",
    type: "uint256"
  }, {
    indexed: false,
    internalType: "bytes32",
    name: "root",
    type: "bytes32"
  }],
  name: "NewHeaderBlock",
  type: "event"
}, {
  inputs: [],
  name: "slash",
  outputs: [],
  stateMutability: "nonpayable",
  type: "function"
}, {
  inputs: [{
    internalType: "bytes",
    name: "data",
    type: "bytes"
  }, {
    internalType: "bytes",
    name: "sigs",
    type: "bytes"
  }],
  name: "submitHeaderBlock",
  outputs: [],
  stateMutability: "nonpayable",
  type: "function"
}, {
  inputs: [],
  name: "getLastChildBlock",
  outputs: [{
    internalType: "uint256",
    name: "",
    type: "uint256"
  }],
  stateMutability: "view",
  type: "function"
}, {
  inputs: [],
  name: "currentHeaderBlock",
  outputs: [{
    internalType: "uint256",
    name: "",
    type: "uint256"
  }],
  stateMutability: "view",
  type: "function"
}];
