// smart-trade-networks/lib/types/factories/ISTN721Manager__factory.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ISTN721Manager__factory = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _ethers = require("ethers");

/* Autogenerated file. Do not edit manually. */

/* tslint:disable */

/* eslint-disable */
var ISTN721Manager__factory = /*#__PURE__*/function () {
  function ISTN721Manager__factory() {
    (0, _classCallCheck2["default"])(this, ISTN721Manager__factory);
  }

  (0, _createClass2["default"])(ISTN721Manager__factory, null, [{
    key: "connect",
    value: function connect(address, signerOrProvider) {
      return new _ethers.Contract(address, _abi, signerOrProvider);
    }
  }]);
  return ISTN721Manager__factory;
}();

exports.ISTN721Manager__factory = ISTN721Manager__factory;
var _abi = [{
  anonymous: false,
  inputs: [{
    indexed: true,
    internalType: "address",
    name: "newController",
    type: "address"
  }],
  name: "AddController",
  type: "event"
}, {
  anonymous: false,
  inputs: [{
    indexed: true,
    internalType: "address",
    name: "stn721Parent",
    type: "address"
  }, {
    indexed: false,
    internalType: "uint256",
    name: "tokenId",
    type: "uint256"
  }, {
    indexed: false,
    internalType: "string",
    name: "name",
    type: "string"
  }, {
    indexed: false,
    internalType: "string",
    name: "symbol",
    type: "string"
  }],
  name: "CreateSTN721Child",
  type: "event"
}, {
  anonymous: false,
  inputs: [{
    indexed: false,
    internalType: "string",
    name: "name",
    type: "string"
  }, {
    indexed: false,
    internalType: "string",
    name: "symbol",
    type: "string"
  }],
  name: "CreateSTN721Parent",
  type: "event"
}, {
  anonymous: false,
  inputs: [{
    indexed: true,
    internalType: "address",
    name: "controller",
    type: "address"
  }],
  name: "RemoveController",
  type: "event"
}, {
  inputs: [{
    internalType: "address",
    name: "_newController",
    type: "address"
  }],
  name: "addController",
  outputs: [],
  stateMutability: "nonpayable",
  type: "function"
}, {
  inputs: [{
    internalType: "address",
    name: "_controller",
    type: "address"
  }],
  name: "removeController",
  outputs: [],
  stateMutability: "nonpayable",
  type: "function"
}, {
  inputs: [{
    internalType: "string",
    name: "name",
    type: "string"
  }, {
    internalType: "string",
    name: "symbol",
    type: "string"
  }],
  name: "createParentSTN721",
  outputs: [{
    internalType: "address",
    name: "",
    type: "address"
  }],
  stateMutability: "nonpayable",
  type: "function"
}, {
  inputs: [{
    internalType: "address",
    name: "stn721Parent",
    type: "address"
  }, {
    internalType: "uint256",
    name: "tokenId",
    type: "uint256"
  }, {
    internalType: "string",
    name: "name",
    type: "string"
  }, {
    internalType: "string",
    name: "symbol",
    type: "string"
  }],
  name: "createChildSTN721",
  outputs: [{
    internalType: "address",
    name: "",
    type: "address"
  }],
  stateMutability: "nonpayable",
  type: "function"
}, {
  inputs: [{
    internalType: "address",
    name: "stn721",
    type: "address"
  }, {
    internalType: "bool",
    name: "divisible",
    type: "bool"
  }, {
    internalType: "uint256[]",
    name: "ids",
    type: "uint256[]"
  }, {
    internalType: "address",
    name: "receiver",
    type: "address"
  }],
  name: "mint",
  outputs: [],
  stateMutability: "nonpayable",
  type: "function"
}, {
  inputs: [{
    internalType: "address",
    name: "stn721Child",
    type: "address"
  }, {
    internalType: "bool",
    name: "divisible",
    type: "bool"
  }, {
    internalType: "uint256[]",
    name: "tokenIds",
    type: "uint256[]"
  }],
  name: "mintChild",
  outputs: [],
  stateMutability: "nonpayable",
  type: "function"
}, {
  inputs: [{
    internalType: "address",
    name: "stn721Parent",
    type: "address"
  }, {
    internalType: "uint256",
    name: "tokenId",
    type: "uint256"
  }],
  name: "getChild",
  outputs: [{
    internalType: "address",
    name: "",
    type: "address"
  }],
  stateMutability: "view",
  type: "function"
}, {
  inputs: [],
  name: "getAll",
  outputs: [{
    internalType: "address[]",
    name: "",
    type: "address[]"
  }],
  stateMutability: "view",
  type: "function"
}];
