// smart-trade-networks/lib/types/factories/Mock__factory.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Mock__factory = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _get2 = _interopRequireDefault(require("@babel/runtime/helpers/get"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _contracts = require("@ethersproject/contracts");

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2["default"])(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2["default"])(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2["default"])(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

var Mock__factory = /*#__PURE__*/function (_ContractFactory) {
  (0, _inherits2["default"])(Mock__factory, _ContractFactory);

  var _super = _createSuper(Mock__factory);

  function Mock__factory(signer) {
    (0, _classCallCheck2["default"])(this, Mock__factory);
    return _super.call(this, _abi, _bytecode, signer);
  }

  (0, _createClass2["default"])(Mock__factory, [{
    key: "deploy",
    value: function deploy(overrides) {
      return (0, _get2["default"])((0, _getPrototypeOf2["default"])(Mock__factory.prototype), "deploy", this).call(this, overrides || {});
    }
  }, {
    key: "getDeployTransaction",
    value: function getDeployTransaction(overrides) {
      return (0, _get2["default"])((0, _getPrototypeOf2["default"])(Mock__factory.prototype), "getDeployTransaction", this).call(this, overrides || {});
    }
  }, {
    key: "attach",
    value: function attach(address) {
      return (0, _get2["default"])((0, _getPrototypeOf2["default"])(Mock__factory.prototype), "attach", this).call(this, address);
    }
  }, {
    key: "connect",
    value: function connect(signer) {
      return (0, _get2["default"])((0, _getPrototypeOf2["default"])(Mock__factory.prototype), "connect", this).call(this, signer);
    }
  }], [{
    key: "connect",
    value: function connect(address, signerOrProvider) {
      return new _contracts.Contract(address, _abi, signerOrProvider);
    }
  }]);
  return Mock__factory;
}(_contracts.ContractFactory);

exports.Mock__factory = Mock__factory;
var _abi = [{
  inputs: [],
  name: "getConstant",
  outputs: [{
    internalType: "uint256",
    name: "",
    type: "uint256"
  }],
  stateMutability: "pure",
  type: "function"
}];
var _bytecode = "0x6080604052348015600f57600080fd5b50608d8061001e6000396000f3fe6080604052348015600f57600080fd5b506004361060285760003560e01c8063f13a38a614602d575b600080fd5b60336047565b604051603e9190604e565b60405180910390f35b6201e24090565b9081526020019056fea26469706673582212201c1f5ff810fc8c8e40291534f98a4b71f2f296a00eaa9249d23b51c025be8ced64736f6c63430008000033";
