// smart-trade-networks/lib/types/factories/Proxy__factory.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Proxy__factory = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _ethers = require("ethers");

/* Autogenerated file. Do not edit manually. */

/* tslint:disable */

/* eslint-disable */
var Proxy__factory = /*#__PURE__*/function () {
  function Proxy__factory() {
    (0, _classCallCheck2["default"])(this, Proxy__factory);
  }

  (0, _createClass2["default"])(Proxy__factory, null, [{
    key: "connect",
    value: function connect(address, signerOrProvider) {
      return new _ethers.Contract(address, _abi, signerOrProvider);
    }
  }]);
  return Proxy__factory;
}();

exports.Proxy__factory = Proxy__factory;
var _abi = [{
  stateMutability: "payable",
  type: "fallback"
}, {
  stateMutability: "payable",
  type: "receive"
}, {
  inputs: [],
  name: "implementation",
  outputs: [{
    internalType: "address",
    name: "",
    type: "address"
  }],
  stateMutability: "view",
  type: "function"
}];
