// smart-trade-networks/lib/controllers/multisig/feeCollectorDeployerController.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _typeof = require("@babel/runtime/helpers/typeof");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.FeeCollectorDeployerController = void 0;

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _ethers = require("ethers");

var _prefabContractFactory = require("../../chain/prefabContractFactory");

var artifact = _interopRequireWildcard(require("../../ethereum/abi/MultisigWallet.json"));

var _encodeCall = require("../../ethereum/encodeCall");

var _deploy = require("../../ethereum/deploy/deploy");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

var FeeCollectorDeployerController = /*#__PURE__*/function () {
  function FeeCollectorDeployerController(feeCollectorAddress, signerOrProvider) {
    (0, _classCallCheck2["default"])(this, FeeCollectorDeployerController);
    (0, _defineProperty2["default"])(this, "_signer", void 0);
    (0, _defineProperty2["default"])(this, "_feeCollectorDeployerAddress", void 0);
    (0, _defineProperty2["default"])(this, "_feeCollectorDeployerContract", void 0);
    this._feeCollectorDeployerAddress = feeCollectorAddress;

    if (signerOrProvider instanceof _ethers.Signer || signerOrProvider instanceof _ethers.ethers.providers.JsonRpcSigner) {
      this._signer = signerOrProvider;
    } else {
      var web3Wrapper = new _ethers.ethers.providers.Web3Provider(signerOrProvider);
      this._signer = web3Wrapper.getSigner();
    }
  }

  (0, _createClass2["default"])(FeeCollectorDeployerController, [{
    key: "_getcontract",
    value: function () {
      var _getcontract2 = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee() {
        var contract;
        return _regenerator["default"].wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                if (!this._feeCollectorDeployerContract) {
                  _context.next = 2;
                  break;
                }

                return _context.abrupt("return", this._feeCollectorDeployerContract);

              case 2:
                _context.prev = 2;
                _context.next = 5;
                return (0, _prefabContractFactory.getFeeCollectorDeployerContract)(this._feeCollectorDeployerAddress, this._signer);

              case 5:
                contract = _context.sent;
                this._feeCollectorDeployerContract = contract;
                return _context.abrupt("return", contract);

              case 10:
                _context.prev = 10;
                _context.t0 = _context["catch"](2);
                throw Error("Failed to get Fee Collector Deployer contract: ".concat(_context.t0));

              case 13:
              case "end":
                return _context.stop();
            }
          }
        }, _callee, this, [[2, 10]]);
      }));

      function _getcontract() {
        return _getcontract2.apply(this, arguments);
      }

      return _getcontract;
    }()
  }, {
    key: "getContract",
    value:
    /**Getters */
    function () {
      var _getContract = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee2() {
        var contract;
        return _regenerator["default"].wrap(function _callee2$(_context2) {
          while (1) {
            switch (_context2.prev = _context2.next) {
              case 0:
                _context2.next = 2;
                return this._getcontract();

              case 2:
                contract = _context2.sent;
                return _context2.abrupt("return", contract);

              case 4:
              case "end":
                return _context2.stop();
            }
          }
        }, _callee2, this);
      }));

      function getContract() {
        return _getContract.apply(this, arguments);
      }

      return getContract;
    }()
  }, {
    key: "getFeeCollectorBytecode",
    value: function () {
      var _getFeeCollectorBytecode = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee3() {
        var erc721Contract;
        return _regenerator["default"].wrap(function _callee3$(_context3) {
          while (1) {
            switch (_context3.prev = _context3.next) {
              case 0:
                _context3.next = 2;
                return this._getcontract();

              case 2:
                erc721Contract = _context3.sent;
                return _context3.abrupt("return", erc721Contract.getFeeCollectorBytecode());

              case 4:
              case "end":
                return _context3.stop();
            }
          }
        }, _callee3, this);
      }));

      function getFeeCollectorBytecode() {
        return _getFeeCollectorBytecode.apply(this, arguments);
      }

      return getFeeCollectorBytecode;
    }()
  }, {
    key: "getProxyBytecode",
    value: function () {
      var _getProxyBytecode = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee4() {
        var erc721Contract;
        return _regenerator["default"].wrap(function _callee4$(_context4) {
          while (1) {
            switch (_context4.prev = _context4.next) {
              case 0:
                _context4.next = 2;
                return this._getcontract();

              case 2:
                erc721Contract = _context4.sent;
                return _context4.abrupt("return", erc721Contract.getProxyBytecode());

              case 4:
              case "end":
                return _context4.stop();
            }
          }
        }, _callee4, this);
      }));

      function getProxyBytecode() {
        return _getProxyBytecode.apply(this, arguments);
      }

      return getProxyBytecode;
    }()
    /** @param bytecode  can be retrieved by calling get___Bytecode()*/

    /** @param _salt is a random number used to create an address */

  }, {
    key: "getAddress",
    value: function () {
      var _getAddress = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee5(bytecode, salt) {
        var erc721Contract;
        return _regenerator["default"].wrap(function _callee5$(_context5) {
          while (1) {
            switch (_context5.prev = _context5.next) {
              case 0:
                _context5.next = 2;
                return this._getcontract();

              case 2:
                erc721Contract = _context5.sent;
                return _context5.abrupt("return", erc721Contract.getAddress(bytecode, salt));

              case 4:
              case "end":
                return _context5.stop();
            }
          }
        }, _callee5, this);
      }));

      function getAddress(_x, _x2) {
        return _getAddress.apply(this, arguments);
      }

      return getAddress;
    }()
  }, {
    key: "getSalt",
    value: function () {
      var _getSalt = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee6(contractAddress) {
        var erc721Contract;
        return _regenerator["default"].wrap(function _callee6$(_context6) {
          while (1) {
            switch (_context6.prev = _context6.next) {
              case 0:
                _context6.next = 2;
                return this._getcontract();

              case 2:
                erc721Contract = _context6.sent;
                return _context6.abrupt("return", erc721Contract.getContractSalt(contractAddress));

              case 4:
              case "end":
                return _context6.stop();
            }
          }
        }, _callee6, this);
      }));

      function getSalt(_x3) {
        return _getSalt.apply(this, arguments);
      }

      return getSalt;
    }()
    /**Setters */

  }, {
    key: "deployPrecomputed",
    value: function () {
      var _deployPrecomputed = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee7(bytecode, salt) {
        var contract;
        return _regenerator["default"].wrap(function _callee7$(_context7) {
          while (1) {
            switch (_context7.prev = _context7.next) {
              case 0:
                _context7.next = 2;
                return this._getcontract();

              case 2:
                contract = _context7.sent;
                return _context7.abrupt("return", contract.functions.deployPrecomputed(bytecode, salt));

              case 4:
              case "end":
                return _context7.stop();
            }
          }
        }, _callee7, this);
      }));

      function deployPrecomputed(_x4, _x5) {
        return _deployPrecomputed.apply(this, arguments);
      }

      return deployPrecomputed;
    }()
  }, {
    key: "setProxyOwner",
    value: function () {
      var _setProxyOwner = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee8(newOwner, proxy) {
        var contract;
        return _regenerator["default"].wrap(function _callee8$(_context8) {
          while (1) {
            switch (_context8.prev = _context8.next) {
              case 0:
                _context8.next = 2;
                return this._getcontract();

              case 2:
                contract = _context8.sent;
                return _context8.abrupt("return", contract.functions.setProxyOwner(newOwner, proxy));

              case 4:
              case "end":
                return _context8.stop();
            }
          }
        }, _callee8, this);
      }));

      function setProxyOwner(_x6, _x7) {
        return _setProxyOwner.apply(this, arguments);
      }

      return setProxyOwner;
    }()
    /** Helpers */

  }], [{
    key: "deployFeeCollectorDeployer",
    value: function () {
      var _deployFeeCollectorDeployer2 = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee9(signer) {
        return _regenerator["default"].wrap(function _callee9$(_context9) {
          while (1) {
            switch (_context9.prev = _context9.next) {
              case 0:
                return _context9.abrupt("return", (0, _deploy.deployFeeCollectorDeployer)(signer));

              case 1:
              case "end":
                return _context9.stop();
            }
          }
        }, _callee9);
      }));

      function deployFeeCollectorDeployer(_x8) {
        return _deployFeeCollectorDeployer2.apply(this, arguments);
      }

      return deployFeeCollectorDeployer;
    }()
  }, {
    key: "encodeCall",
    value: function encodeCall(functionName, args) {
      return (0, _encodeCall.encode)(artifact, functionName, args);
    }
  }, {
    key: "decodeCall",
    value: function decodeCall(data) {
      return (0, _encodeCall.decode)(artifact, data);
    }
  }, {
    key: "getBytecode",
    value: function getBytecode() {
      return artifact.bytecode;
    }
  }]);
  return FeeCollectorDeployerController;
}();

exports.FeeCollectorDeployerController = FeeCollectorDeployerController;
