// smart-trade-networks/lib/controllers/multisig/feeCollectorController.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _typeof = require("@babel/runtime/helpers/typeof");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.FeeCollectorController = void 0;

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _ethers = require("ethers");

var _prefabContractFactory = require("../../chain/prefabContractFactory");

var artifact = _interopRequireWildcard(require("../../ethereum/abi/FeeCollector.json"));

var _encodeCall = require("../../ethereum/encodeCall");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

var FeeCollectorController = /*#__PURE__*/function () {
  function FeeCollectorController(feeCollectorAddress, signerOrProvider) {
    (0, _classCallCheck2["default"])(this, FeeCollectorController);
    (0, _defineProperty2["default"])(this, "_signer", void 0);
    (0, _defineProperty2["default"])(this, "_feeCollectorAddress", void 0);
    (0, _defineProperty2["default"])(this, "_feeCollectorContract", void 0);
    this._feeCollectorAddress = feeCollectorAddress;

    if (signerOrProvider instanceof _ethers.Signer || signerOrProvider instanceof _ethers.ethers.providers.JsonRpcSigner) {
      this._signer = signerOrProvider;
    } else {
      var web3Wrapper = new _ethers.ethers.providers.Web3Provider(signerOrProvider);
      this._signer = web3Wrapper.getSigner();
    }
  }

  (0, _createClass2["default"])(FeeCollectorController, [{
    key: "_getcontract",
    value: function () {
      var _getcontract2 = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee() {
        var contract;
        return _regenerator["default"].wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                if (!this._feeCollectorContract) {
                  _context.next = 2;
                  break;
                }

                return _context.abrupt("return", this._feeCollectorContract);

              case 2:
                _context.prev = 2;
                _context.next = 5;
                return (0, _prefabContractFactory.getFeeCollectorContract)(this._feeCollectorAddress, this._signer);

              case 5:
                contract = _context.sent;
                this._feeCollectorContract = contract;
                return _context.abrupt("return", contract);

              case 10:
                _context.prev = 10;
                _context.t0 = _context["catch"](2);
                throw Error("Failed to get Fee Collector contract: ".concat(_context.t0));

              case 13:
              case "end":
                return _context.stop();
            }
          }
        }, _callee, this, [[2, 10]]);
      }));

      function _getcontract() {
        return _getcontract2.apply(this, arguments);
      }

      return _getcontract;
    }()
    /**Getters */

  }, {
    key: "getContract",
    value: function () {
      var _getContract = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee2() {
        var contract;
        return _regenerator["default"].wrap(function _callee2$(_context2) {
          while (1) {
            switch (_context2.prev = _context2.next) {
              case 0:
                _context2.next = 2;
                return this._getcontract();

              case 2:
                contract = _context2.sent;
                return _context2.abrupt("return", contract);

              case 4:
              case "end":
                return _context2.stop();
            }
          }
        }, _callee2, this);
      }));

      function getContract() {
        return _getContract.apply(this, arguments);
      }

      return getContract;
    }()
  }, {
    key: "getFeeByMultisig",
    value: function () {
      var _getFeeByMultisig = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee3(address) {
        var contract;
        return _regenerator["default"].wrap(function _callee3$(_context3) {
          while (1) {
            switch (_context3.prev = _context3.next) {
              case 0:
                _context3.next = 2;
                return this._getcontract();

              case 2:
                contract = _context3.sent;
                return _context3.abrupt("return", contract.getFeeByMultisig(address));

              case 4:
              case "end":
                return _context3.stop();
            }
          }
        }, _callee3, this);
      }));

      function getFeeByMultisig(_x) {
        return _getFeeByMultisig.apply(this, arguments);
      }

      return getFeeByMultisig;
    }()
  }, {
    key: "getFee",
    value: function () {
      var _getFee = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee4() {
        var contract;
        return _regenerator["default"].wrap(function _callee4$(_context4) {
          while (1) {
            switch (_context4.prev = _context4.next) {
              case 0:
                _context4.next = 2;
                return this._getcontract();

              case 2:
                contract = _context4.sent;
                return _context4.abrupt("return", contract.getFee());

              case 4:
              case "end":
                return _context4.stop();
            }
          }
        }, _callee4, this);
      }));

      function getFee() {
        return _getFee.apply(this, arguments);
      }

      return getFee;
    }()
  }, {
    key: "getFeeCollector",
    value: function () {
      var _getFeeCollector = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee5() {
        var contract;
        return _regenerator["default"].wrap(function _callee5$(_context5) {
          while (1) {
            switch (_context5.prev = _context5.next) {
              case 0:
                _context5.next = 2;
                return this._getcontract();

              case 2:
                contract = _context5.sent;
                return _context5.abrupt("return", contract.getFeeCollector());

              case 4:
              case "end":
                return _context5.stop();
            }
          }
        }, _callee5, this);
      }));

      function getFeeCollector() {
        return _getFeeCollector.apply(this, arguments);
      }

      return getFeeCollector;
    }()
  }, {
    key: "isWhitelisted",
    value: function () {
      var _isWhitelisted = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee6(address) {
        var contract;
        return _regenerator["default"].wrap(function _callee6$(_context6) {
          while (1) {
            switch (_context6.prev = _context6.next) {
              case 0:
                _context6.next = 2;
                return this._getcontract();

              case 2:
                contract = _context6.sent;
                return _context6.abrupt("return", contract.isWhitelisted(address));

              case 4:
              case "end":
                return _context6.stop();
            }
          }
        }, _callee6, this);
      }));

      function isWhitelisted(_x2) {
        return _isWhitelisted.apply(this, arguments);
      }

      return isWhitelisted;
    }()
    /**Setters */

  }, {
    key: "setFeeCollector",
    value: function () {
      var _setFeeCollector = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee7(address) {
        var contract;
        return _regenerator["default"].wrap(function _callee7$(_context7) {
          while (1) {
            switch (_context7.prev = _context7.next) {
              case 0:
                _context7.next = 2;
                return this._getcontract();

              case 2:
                contract = _context7.sent;
                return _context7.abrupt("return", contract.functions.setFeeCollector(address));

              case 4:
              case "end":
                return _context7.stop();
            }
          }
        }, _callee7, this);
      }));

      function setFeeCollector(_x3) {
        return _setFeeCollector.apply(this, arguments);
      }

      return setFeeCollector;
    }()
  }, {
    key: "setFee",
    value: function () {
      var _setFee = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee8(newFee) {
        var contract;
        return _regenerator["default"].wrap(function _callee8$(_context8) {
          while (1) {
            switch (_context8.prev = _context8.next) {
              case 0:
                _context8.next = 2;
                return this._getcontract();

              case 2:
                contract = _context8.sent;
                return _context8.abrupt("return", contract.functions.setFee(newFee));

              case 4:
              case "end":
                return _context8.stop();
            }
          }
        }, _callee8, this);
      }));

      function setFee(_x4) {
        return _setFee.apply(this, arguments);
      }

      return setFee;
    }()
  }, {
    key: "setFeeByMultisig",
    value: function () {
      var _setFeeByMultisig = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee9(address, newFee) {
        var contract;
        return _regenerator["default"].wrap(function _callee9$(_context9) {
          while (1) {
            switch (_context9.prev = _context9.next) {
              case 0:
                _context9.next = 2;
                return this._getcontract();

              case 2:
                contract = _context9.sent;
                return _context9.abrupt("return", contract.functions.setFeeByMultisig(address, newFee));

              case 4:
              case "end":
                return _context9.stop();
            }
          }
        }, _callee9, this);
      }));

      function setFeeByMultisig(_x5, _x6) {
        return _setFeeByMultisig.apply(this, arguments);
      }

      return setFeeByMultisig;
    }()
  }, {
    key: "whitelistMultisig",
    value: function () {
      var _whitelistMultisig = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee10(address, allow) {
        var contract;
        return _regenerator["default"].wrap(function _callee10$(_context10) {
          while (1) {
            switch (_context10.prev = _context10.next) {
              case 0:
                _context10.next = 2;
                return this._getcontract();

              case 2:
                contract = _context10.sent;
                return _context10.abrupt("return", contract.functions.whitelistMultisig(address, allow));

              case 4:
              case "end":
                return _context10.stop();
            }
          }
        }, _callee10, this);
      }));

      function whitelistMultisig(_x7, _x8) {
        return _whitelistMultisig.apply(this, arguments);
      }

      return whitelistMultisig;
    }()
  }, {
    key: "withdraw",
    value: function () {
      var _withdraw = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee11(tokenAddress) {
        var contract;
        return _regenerator["default"].wrap(function _callee11$(_context11) {
          while (1) {
            switch (_context11.prev = _context11.next) {
              case 0:
                _context11.next = 2;
                return this._getcontract();

              case 2:
                contract = _context11.sent;
                return _context11.abrupt("return", contract.functions.withdraw(tokenAddress));

              case 4:
              case "end":
                return _context11.stop();
            }
          }
        }, _callee11, this);
      }));

      function withdraw(_x9) {
        return _withdraw.apply(this, arguments);
      }

      return withdraw;
    }()
    /** Helpers */

  }], [{
    key: "encodeCall",
    value: function encodeCall(functionName, args) {
      return (0, _encodeCall.encode)(artifact, functionName, args);
    }
  }, {
    key: "decodeCall",
    value: function decodeCall(data) {
      return (0, _encodeCall.decode)(artifact, data);
    }
  }]);
  return FeeCollectorController;
}();

exports.FeeCollectorController = FeeCollectorController;
