// smart-trade-networks/lib/controllers/multisig/mulsitigDeployerController.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _typeof = require("@babel/runtime/helpers/typeof");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MultisigDeployerController = void 0;

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _ethers = require("ethers");

var _prefabContractFactory = require("../../chain/prefabContractFactory");

var artifact = _interopRequireWildcard(require("../../ethereum/abi/MultisigDeployer.json"));

var _encodeCall = require("../../ethereum/encodeCall");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

var MultisigDeployerController = /*#__PURE__*/function () {
  function MultisigDeployerController(multisigDeployerAddress, signerOrProvider) {
    (0, _classCallCheck2["default"])(this, MultisigDeployerController);
    (0, _defineProperty2["default"])(this, "_signer", void 0);
    (0, _defineProperty2["default"])(this, "_multisigDeployerAddress", void 0);
    (0, _defineProperty2["default"])(this, "_multisigDeployerContract", void 0);
    this._multisigDeployerAddress = multisigDeployerAddress;

    if (signerOrProvider instanceof _ethers.Signer || signerOrProvider instanceof _ethers.ethers.providers.JsonRpcSigner) {
      this._signer = signerOrProvider;
    } else {
      var web3Wrapper = new _ethers.ethers.providers.Web3Provider(signerOrProvider);
      this._signer = web3Wrapper.getSigner();
    }
  }

  (0, _createClass2["default"])(MultisigDeployerController, [{
    key: "_getContract",
    value: function () {
      var _getContract2 = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee() {
        var contract;
        return _regenerator["default"].wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                if (!this._multisigDeployerContract) {
                  _context.next = 2;
                  break;
                }

                return _context.abrupt("return", this._multisigDeployerContract);

              case 2:
                _context.prev = 2;
                _context.next = 5;
                return (0, _prefabContractFactory.getMultisigDeployerContract)(this._multisigDeployerAddress, this._signer);

              case 5:
                contract = _context.sent;
                this._multisigDeployerContract = contract;
                return _context.abrupt("return", contract);

              case 10:
                _context.prev = 10;
                _context.t0 = _context["catch"](2);
                throw Error("Failed to get Multisig contract: ".concat(_context.t0));

              case 13:
              case "end":
                return _context.stop();
            }
          }
        }, _callee, this, [[2, 10]]);
      }));

      function _getContract() {
        return _getContract2.apply(this, arguments);
      }

      return _getContract;
    }()
    /**Getters */

  }, {
    key: "getContract",
    value: function () {
      var _getContract3 = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee2() {
        var contract;
        return _regenerator["default"].wrap(function _callee2$(_context2) {
          while (1) {
            switch (_context2.prev = _context2.next) {
              case 0:
                _context2.next = 2;
                return this._getContract();

              case 2:
                contract = _context2.sent;
                return _context2.abrupt("return", contract);

              case 4:
              case "end":
                return _context2.stop();
            }
          }
        }, _callee2, this);
      }));

      function getContract() {
        return _getContract3.apply(this, arguments);
      }

      return getContract;
    }()
  }, {
    key: "getAll",
    value: function () {
      var _getAll = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee3() {
        var contract;
        return _regenerator["default"].wrap(function _callee3$(_context3) {
          while (1) {
            switch (_context3.prev = _context3.next) {
              case 0:
                _context3.next = 2;
                return this._getContract();

              case 2:
                contract = _context3.sent;
                return _context3.abrupt("return", contract.getAll());

              case 4:
              case "end":
                return _context3.stop();
            }
          }
        }, _callee3, this);
      }));

      function getAll() {
        return _getAll.apply(this, arguments);
      }

      return getAll;
    }()
  }, {
    key: "isMultisigAdded",
    value: function () {
      var _isMultisigAdded = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee4(multisigAddress) {
        var contract;
        return _regenerator["default"].wrap(function _callee4$(_context4) {
          while (1) {
            switch (_context4.prev = _context4.next) {
              case 0:
                _context4.next = 2;
                return this._getContract();

              case 2:
                contract = _context4.sent;
                return _context4.abrupt("return", contract.isMultisigAdded(multisigAddress));

              case 4:
              case "end":
                return _context4.stop();
            }
          }
        }, _callee4, this);
      }));

      function isMultisigAdded(_x) {
        return _isMultisigAdded.apply(this, arguments);
      }

      return isMultisigAdded;
    }()
  }, {
    key: "getFeeCollector",
    value: function () {
      var _getFeeCollector = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee5() {
        var contract;
        return _regenerator["default"].wrap(function _callee5$(_context5) {
          while (1) {
            switch (_context5.prev = _context5.next) {
              case 0:
                _context5.next = 2;
                return this._getContract();

              case 2:
                contract = _context5.sent;
                return _context5.abrupt("return", contract.feeCollector());

              case 4:
              case "end":
                return _context5.stop();
            }
          }
        }, _callee5, this);
      }));

      function getFeeCollector() {
        return _getFeeCollector.apply(this, arguments);
      }

      return getFeeCollector;
    }()
  }, {
    key: "getPrecomputedAddress",
    value: function () {
      var _getPrecomputedAddress = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee6(bytecode, salt) {
        var contract;
        return _regenerator["default"].wrap(function _callee6$(_context6) {
          while (1) {
            switch (_context6.prev = _context6.next) {
              case 0:
                _context6.next = 2;
                return this._getContract();

              case 2:
                contract = _context6.sent;
                return _context6.abrupt("return", contract.getAddress(bytecode, salt));

              case 4:
              case "end":
                return _context6.stop();
            }
          }
        }, _callee6, this);
      }));

      function getPrecomputedAddress(_x2, _x3) {
        return _getPrecomputedAddress.apply(this, arguments);
      }

      return getPrecomputedAddress;
    }()
  }, {
    key: "getBytecode",
    value: function () {
      var _getBytecode = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee7(owners, requried) {
        var contract;
        return _regenerator["default"].wrap(function _callee7$(_context7) {
          while (1) {
            switch (_context7.prev = _context7.next) {
              case 0:
                _context7.next = 2;
                return this._getContract();

              case 2:
                contract = _context7.sent;
                return _context7.abrupt("return", contract.getBytecode(owners, requried));

              case 4:
              case "end":
                return _context7.stop();
            }
          }
        }, _callee7, this);
      }));

      function getBytecode(_x4, _x5) {
        return _getBytecode.apply(this, arguments);
      }

      return getBytecode;
    }()
  }, {
    key: "getContractSalt",
    value: function () {
      var _getContractSalt = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee8(address) {
        var contract;
        return _regenerator["default"].wrap(function _callee8$(_context8) {
          while (1) {
            switch (_context8.prev = _context8.next) {
              case 0:
                _context8.next = 2;
                return this._getContract();

              case 2:
                contract = _context8.sent;
                return _context8.abrupt("return", contract.getContractSalt(address));

              case 4:
              case "end":
                return _context8.stop();
            }
          }
        }, _callee8, this);
      }));

      function getContractSalt(_x6) {
        return _getContractSalt.apply(this, arguments);
      }

      return getContractSalt;
    }()
    /**Setters */

  }, {
    key: "setFeeCollector",
    value: function () {
      var _setFeeCollector = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee9(feeCollector) {
        var contract;
        return _regenerator["default"].wrap(function _callee9$(_context9) {
          while (1) {
            switch (_context9.prev = _context9.next) {
              case 0:
                _context9.next = 2;
                return this._getContract();

              case 2:
                contract = _context9.sent;
                return _context9.abrupt("return", contract.functions.setFeeCollector(feeCollector));

              case 4:
              case "end":
                return _context9.stop();
            }
          }
        }, _callee9, this);
      }));

      function setFeeCollector(_x7) {
        return _setFeeCollector.apply(this, arguments);
      }

      return setFeeCollector;
    }()
  }, {
    key: "deployMultisigPrecomputed",
    value: function () {
      var _deployMultisigPrecomputed = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee10(bytecode, salt) {
        var contract;
        return _regenerator["default"].wrap(function _callee10$(_context10) {
          while (1) {
            switch (_context10.prev = _context10.next) {
              case 0:
                _context10.next = 2;
                return this._getContract();

              case 2:
                contract = _context10.sent;
                return _context10.abrupt("return", contract.functions.deployMultisig(bytecode, salt));

              case 4:
              case "end":
                return _context10.stop();
            }
          }
        }, _callee10, this);
      }));

      function deployMultisigPrecomputed(_x8, _x9) {
        return _deployMultisigPrecomputed.apply(this, arguments);
      }

      return deployMultisigPrecomputed;
    }()
  }, {
    key: "transferOwnership",
    value: function () {
      var _transferOwnership = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee11(newOwner) {
        var contract;
        return _regenerator["default"].wrap(function _callee11$(_context11) {
          while (1) {
            switch (_context11.prev = _context11.next) {
              case 0:
                _context11.next = 2;
                return this._getContract();

              case 2:
                contract = _context11.sent;
                return _context11.abrupt("return", contract.functions.transferOwnership(newOwner));

              case 4:
              case "end":
                return _context11.stop();
            }
          }
        }, _callee11, this);
      }));

      function transferOwnership(_x10) {
        return _transferOwnership.apply(this, arguments);
      }

      return transferOwnership;
    }()
    /** Helpers */

  }], [{
    key: "encodeCall",
    value: function encodeCall(functionName, args) {
      return (0, _encodeCall.encode)(artifact, functionName, args);
    }
  }, {
    key: "decodeCall",
    value: function decodeCall(data) {
      return (0, _encodeCall.decode)(artifact, data);
    }
  }]);
  return MultisigDeployerController;
}();

exports.MultisigDeployerController = MultisigDeployerController;
