// smart-trade-networks/lib/controllers/multisig/multisigController.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _typeof = require("@babel/runtime/helpers/typeof");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MultisigController = void 0;

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _ethers = require("ethers");

var _prefabContractFactory = require("../../chain/prefabContractFactory");

var _deploy = require("../../ethereum/deploy/deploy");

var artifact = _interopRequireWildcard(require("../../ethereum/abi/MultisigWallet.json"));

var _encodeCall = require("../../ethereum/encodeCall");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

var MultisigController = /*#__PURE__*/function () {
  function MultisigController(multisigAddress, signerOrProvider) {
    (0, _classCallCheck2["default"])(this, MultisigController);
    (0, _defineProperty2["default"])(this, "_signer", void 0);
    (0, _defineProperty2["default"])(this, "_multisigAddress", void 0);
    (0, _defineProperty2["default"])(this, "_multisigContract", void 0);
    this._multisigAddress = multisigAddress;

    if (signerOrProvider instanceof _ethers.Signer || signerOrProvider instanceof _ethers.ethers.providers.JsonRpcSigner) {
      this._signer = signerOrProvider;
    } else {
      var web3Wrapper = new _ethers.ethers.providers.Web3Provider(signerOrProvider);
      this._signer = web3Wrapper.getSigner();
    }
  }

  (0, _createClass2["default"])(MultisigController, [{
    key: "_getMultisigContract",
    value: function () {
      var _getMultisigContract2 = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee() {
        var contract;
        return _regenerator["default"].wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                if (!this._multisigContract) {
                  _context.next = 2;
                  break;
                }

                return _context.abrupt("return", this._multisigContract);

              case 2:
                _context.prev = 2;
                _context.next = 5;
                return (0, _prefabContractFactory.getMultisigContract)(this._multisigAddress, this._signer);

              case 5:
                contract = _context.sent;
                this._multisigContract = contract;
                return _context.abrupt("return", contract);

              case 10:
                _context.prev = 10;
                _context.t0 = _context["catch"](2);
                throw Error("Failed to get Multisig contract: ".concat(_context.t0));

              case 13:
              case "end":
                return _context.stop();
            }
          }
        }, _callee, this, [[2, 10]]);
      }));

      function _getMultisigContract() {
        return _getMultisigContract2.apply(this, arguments);
      }

      return _getMultisigContract;
    }()
    /**Getters */

  }, {
    key: "getContract",
    value: function () {
      var _getContract = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee2() {
        var multisigContract;
        return _regenerator["default"].wrap(function _callee2$(_context2) {
          while (1) {
            switch (_context2.prev = _context2.next) {
              case 0:
                _context2.next = 2;
                return this._getMultisigContract();

              case 2:
                multisigContract = _context2.sent;
                return _context2.abrupt("return", multisigContract);

              case 4:
              case "end":
                return _context2.stop();
            }
          }
        }, _callee2, this);
      }));

      function getContract() {
        return _getContract.apply(this, arguments);
      }

      return getContract;
    }()
  }, {
    key: "isConfirmed",
    value: function () {
      var _isConfirmed = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee3(transactionId) {
        var multisigContract;
        return _regenerator["default"].wrap(function _callee3$(_context3) {
          while (1) {
            switch (_context3.prev = _context3.next) {
              case 0:
                _context3.next = 2;
                return this._getMultisigContract();

              case 2:
                multisigContract = _context3.sent;
                return _context3.abrupt("return", multisigContract.isConfirmed(transactionId));

              case 4:
              case "end":
                return _context3.stop();
            }
          }
        }, _callee3, this);
      }));

      function isConfirmed(_x) {
        return _isConfirmed.apply(this, arguments);
      }

      return isConfirmed;
    }()
  }, {
    key: "getConfirmationCount",
    value: function () {
      var _getConfirmationCount = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee4(transactionId) {
        var multisigContract;
        return _regenerator["default"].wrap(function _callee4$(_context4) {
          while (1) {
            switch (_context4.prev = _context4.next) {
              case 0:
                _context4.next = 2;
                return this._getMultisigContract();

              case 2:
                multisigContract = _context4.sent;
                return _context4.abrupt("return", multisigContract.getConfirmationCount(transactionId));

              case 4:
              case "end":
                return _context4.stop();
            }
          }
        }, _callee4, this);
      }));

      function getConfirmationCount(_x2) {
        return _getConfirmationCount.apply(this, arguments);
      }

      return getConfirmationCount;
    }()
  }, {
    key: "getTransactionCount",
    value: function () {
      var _getTransactionCount = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee5(pending, executed) {
        var multisigContract;
        return _regenerator["default"].wrap(function _callee5$(_context5) {
          while (1) {
            switch (_context5.prev = _context5.next) {
              case 0:
                _context5.next = 2;
                return this._getMultisigContract();

              case 2:
                multisigContract = _context5.sent;
                return _context5.abrupt("return", multisigContract.getTransactionCount(pending, executed));

              case 4:
              case "end":
                return _context5.stop();
            }
          }
        }, _callee5, this);
      }));

      function getTransactionCount(_x3, _x4) {
        return _getTransactionCount.apply(this, arguments);
      }

      return getTransactionCount;
    }()
  }, {
    key: "getConfirmations",
    value: function () {
      var _getConfirmations = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee6(transactionId) {
        var multisigContract;
        return _regenerator["default"].wrap(function _callee6$(_context6) {
          while (1) {
            switch (_context6.prev = _context6.next) {
              case 0:
                _context6.next = 2;
                return this._getMultisigContract();

              case 2:
                multisigContract = _context6.sent;
                return _context6.abrupt("return", multisigContract.getConfirmations(transactionId));

              case 4:
              case "end":
                return _context6.stop();
            }
          }
        }, _callee6, this);
      }));

      function getConfirmations(_x5) {
        return _getConfirmations.apply(this, arguments);
      }

      return getConfirmations;
    }()
  }, {
    key: "getTransactionIds",
    value: function () {
      var _getTransactionIds = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee7(from, to, pending, executed) {
        var multisigContract;
        return _regenerator["default"].wrap(function _callee7$(_context7) {
          while (1) {
            switch (_context7.prev = _context7.next) {
              case 0:
                _context7.next = 2;
                return this._getMultisigContract();

              case 2:
                multisigContract = _context7.sent;
                return _context7.abrupt("return", multisigContract.getTransactionIds(from, to, pending, executed));

              case 4:
              case "end":
                return _context7.stop();
            }
          }
        }, _callee7, this);
      }));

      function getTransactionIds(_x6, _x7, _x8, _x9) {
        return _getTransactionIds.apply(this, arguments);
      }

      return getTransactionIds;
    }()
  }, {
    key: "getOwners",
    value: function () {
      var _getOwners = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee8() {
        var multisigContract;
        return _regenerator["default"].wrap(function _callee8$(_context8) {
          while (1) {
            switch (_context8.prev = _context8.next) {
              case 0:
                _context8.next = 2;
                return this._getMultisigContract();

              case 2:
                multisigContract = _context8.sent;
                return _context8.abrupt("return", multisigContract.getOwners());

              case 4:
              case "end":
                return _context8.stop();
            }
          }
        }, _callee8, this);
      }));

      function getOwners() {
        return _getOwners.apply(this, arguments);
      }

      return getOwners;
    }()
  }, {
    key: "getTransaction",
    value: function () {
      var _getTransaction = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee9(transactionId) {
        var multisigContract;
        return _regenerator["default"].wrap(function _callee9$(_context9) {
          while (1) {
            switch (_context9.prev = _context9.next) {
              case 0:
                _context9.next = 2;
                return this._getMultisigContract();

              case 2:
                multisigContract = _context9.sent;
                return _context9.abrupt("return", multisigContract.getTransaction(transactionId));

              case 4:
              case "end":
                return _context9.stop();
            }
          }
        }, _callee9, this);
      }));

      function getTransaction(_x10) {
        return _getTransaction.apply(this, arguments);
      }

      return getTransaction;
    }()
  }, {
    key: "getResourceActionsContract",
    value: function () {
      var _getResourceActionsContract = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee10() {
        var multisigContract;
        return _regenerator["default"].wrap(function _callee10$(_context10) {
          while (1) {
            switch (_context10.prev = _context10.next) {
              case 0:
                _context10.next = 2;
                return this._getMultisigContract();

              case 2:
                multisigContract = _context10.sent;
                return _context10.abrupt("return", multisigContract.getResourceActionsContract());

              case 4:
              case "end":
                return _context10.stop();
            }
          }
        }, _callee10, this);
      }));

      function getResourceActionsContract() {
        return _getResourceActionsContract.apply(this, arguments);
      }

      return getResourceActionsContract;
    }()
  }, {
    key: "getRequiredSignatures",
    value: function () {
      var _getRequiredSignatures = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee11() {
        var multisigContract;
        return _regenerator["default"].wrap(function _callee11$(_context11) {
          while (1) {
            switch (_context11.prev = _context11.next) {
              case 0:
                _context11.next = 2;
                return this._getMultisigContract();

              case 2:
                multisigContract = _context11.sent;
                return _context11.abrupt("return", multisigContract.required());

              case 4:
              case "end":
                return _context11.stop();
            }
          }
        }, _callee11, this);
      }));

      function getRequiredSignatures() {
        return _getRequiredSignatures.apply(this, arguments);
      }

      return getRequiredSignatures;
    }()
  }, {
    key: "isOwner",
    value: function () {
      var _isOwner = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee12(address) {
        var multisigContract;
        return _regenerator["default"].wrap(function _callee12$(_context12) {
          while (1) {
            switch (_context12.prev = _context12.next) {
              case 0:
                _context12.next = 2;
                return this._getMultisigContract();

              case 2:
                multisigContract = _context12.sent;
                return _context12.abrupt("return", multisigContract.isOwner(address));

              case 4:
              case "end":
                return _context12.stop();
            }
          }
        }, _callee12, this);
      }));

      function isOwner(_x11) {
        return _isOwner.apply(this, arguments);
      }

      return isOwner;
    }()
    /**Setters */

  }, {
    key: "submitTransaction",
    value: function () {
      var _submitTransaction = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee13(destination, value, data, description, overrides) {
        var multisigContract;
        return _regenerator["default"].wrap(function _callee13$(_context13) {
          while (1) {
            switch (_context13.prev = _context13.next) {
              case 0:
                _context13.next = 2;
                return this._getMultisigContract();

              case 2:
                multisigContract = _context13.sent;
                return _context13.abrupt("return", multisigContract.functions.submitTransaction(destination, value, data, description, overrides));

              case 4:
              case "end":
                return _context13.stop();
            }
          }
        }, _callee13, this);
      }));

      function submitTransaction(_x12, _x13, _x14, _x15, _x16) {
        return _submitTransaction.apply(this, arguments);
      }

      return submitTransaction;
    }()
  }, {
    key: "confirmTransaction",
    value: function () {
      var _confirmTransaction = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee14(transactionId, overrides) {
        var multisigContract;
        return _regenerator["default"].wrap(function _callee14$(_context14) {
          while (1) {
            switch (_context14.prev = _context14.next) {
              case 0:
                _context14.next = 2;
                return this._getMultisigContract();

              case 2:
                multisigContract = _context14.sent;
                return _context14.abrupt("return", multisigContract.functions.confirmTransaction(transactionId, overrides));

              case 4:
              case "end":
                return _context14.stop();
            }
          }
        }, _callee14, this);
      }));

      function confirmTransaction(_x17, _x18) {
        return _confirmTransaction.apply(this, arguments);
      }

      return confirmTransaction;
    }()
  }, {
    key: "revokeConfirmation",
    value: function () {
      var _revokeConfirmation = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee15(transactionId) {
        var multisigContract;
        return _regenerator["default"].wrap(function _callee15$(_context15) {
          while (1) {
            switch (_context15.prev = _context15.next) {
              case 0:
                _context15.next = 2;
                return this._getMultisigContract();

              case 2:
                multisigContract = _context15.sent;
                return _context15.abrupt("return", multisigContract.functions.revokeConfirmation(transactionId));

              case 4:
              case "end":
                return _context15.stop();
            }
          }
        }, _callee15, this);
      }));

      function revokeConfirmation(_x19) {
        return _revokeConfirmation.apply(this, arguments);
      }

      return revokeConfirmation;
    }()
  }, {
    key: "executeTransaction",
    value: function () {
      var _executeTransaction = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee16(transactionId) {
        var multisigContract;
        return _regenerator["default"].wrap(function _callee16$(_context16) {
          while (1) {
            switch (_context16.prev = _context16.next) {
              case 0:
                _context16.next = 2;
                return this._getMultisigContract();

              case 2:
                multisigContract = _context16.sent;
                return _context16.abrupt("return", multisigContract.functions.executeTransaction(transactionId));

              case 4:
              case "end":
                return _context16.stop();
            }
          }
        }, _callee16, this);
      }));

      function executeTransaction(_x20) {
        return _executeTransaction.apply(this, arguments);
      }

      return executeTransaction;
    }()
    /** Helpers */

  }], [{
    key: "deployMultisigContract",
    value: function () {
      var _deployMultisigContract = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee17(signer, owners, requirement) {
        return _regenerator["default"].wrap(function _callee17$(_context17) {
          while (1) {
            switch (_context17.prev = _context17.next) {
              case 0:
                return _context17.abrupt("return", (0, _deploy.deployMultisig)(signer, owners, requirement));

              case 1:
              case "end":
                return _context17.stop();
            }
          }
        }, _callee17);
      }));

      function deployMultisigContract(_x21, _x22, _x23) {
        return _deployMultisigContract.apply(this, arguments);
      }

      return deployMultisigContract;
    }()
  }, {
    key: "encodeCall",
    value: function encodeCall(functionName, args) {
      return (0, _encodeCall.encode)(artifact, functionName, args);
    }
  }, {
    key: "decodeCall",
    value: function decodeCall(data) {
      return (0, _encodeCall.decode)(artifact, data);
    }
  }]);
  return MultisigController;
}();

exports.MultisigController = MultisigController;
