// smart-trade-networks/lib/controllers/multisig/multisigENSController.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _typeof = require("@babel/runtime/helpers/typeof");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MultisigENSController = void 0;

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _ethers = require("ethers");

var _prefabContractFactory = require("../../chain/prefabContractFactory");

var artifact = _interopRequireWildcard(require("../../ethereum/abi/MultisigENS.json"));

var _encodeCall = require("../../ethereum/encodeCall");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

var MultisigENSController = /*#__PURE__*/function () {
  function MultisigENSController(multisigAddress, signerOrProvider) {
    (0, _classCallCheck2["default"])(this, MultisigENSController);
    (0, _defineProperty2["default"])(this, "_signer", void 0);
    (0, _defineProperty2["default"])(this, "_multisigENSAddress", void 0);
    (0, _defineProperty2["default"])(this, "_multisigENSContract", void 0);
    this._multisigENSAddress = multisigAddress;

    if (signerOrProvider instanceof _ethers.Signer || signerOrProvider instanceof _ethers.ethers.providers.JsonRpcSigner) {
      this._signer = signerOrProvider;
    } else {
      var web3Wrapper = new _ethers.ethers.providers.Web3Provider(signerOrProvider);
      this._signer = web3Wrapper.getSigner();
    }
  }

  (0, _createClass2["default"])(MultisigENSController, [{
    key: "_getMultisigENSContract",
    value: function () {
      var _getMultisigENSContract2 = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee() {
        var contract;
        return _regenerator["default"].wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                if (!this._multisigENSContract) {
                  _context.next = 2;
                  break;
                }

                return _context.abrupt("return", this._multisigENSContract);

              case 2:
                _context.prev = 2;
                _context.next = 5;
                return (0, _prefabContractFactory.getMultisigENSContract)(this._multisigENSAddress, this._signer);

              case 5:
                contract = _context.sent;
                this._multisigENSContract = contract;
                return _context.abrupt("return", contract);

              case 10:
                _context.prev = 10;
                _context.t0 = _context["catch"](2);
                throw Error("Failed to get Multisig contract: ".concat(_context.t0));

              case 13:
              case "end":
                return _context.stop();
            }
          }
        }, _callee, this, [[2, 10]]);
      }));

      function _getMultisigENSContract() {
        return _getMultisigENSContract2.apply(this, arguments);
      }

      return _getMultisigENSContract;
    }()
    /**Getters */

  }, {
    key: "getContract",
    value: function () {
      var _getContract = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee2() {
        var contract;
        return _regenerator["default"].wrap(function _callee2$(_context2) {
          while (1) {
            switch (_context2.prev = _context2.next) {
              case 0:
                _context2.next = 2;
                return this._getMultisigENSContract();

              case 2:
                contract = _context2.sent;
                return _context2.abrupt("return", contract);

              case 4:
              case "end":
                return _context2.stop();
            }
          }
        }, _callee2, this);
      }));

      function getContract() {
        return _getContract.apply(this, arguments);
      }

      return getContract;
    }()
  }, {
    key: "getMultisigDeployerAddress",
    value: function () {
      var _getMultisigDeployerAddress = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee3() {
        var contract;
        return _regenerator["default"].wrap(function _callee3$(_context3) {
          while (1) {
            switch (_context3.prev = _context3.next) {
              case 0:
                _context3.next = 2;
                return this._getMultisigENSContract();

              case 2:
                contract = _context3.sent;
                return _context3.abrupt("return", contract.getMultisigDeployerAddress());

              case 4:
              case "end":
                return _context3.stop();
            }
          }
        }, _callee3, this);
      }));

      function getMultisigDeployerAddress() {
        return _getMultisigDeployerAddress.apply(this, arguments);
      }

      return getMultisigDeployerAddress;
    }()
  }, {
    key: "getMultisigName",
    value: function () {
      var _getMultisigName = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee4(msigAddress) {
        var contract;
        return _regenerator["default"].wrap(function _callee4$(_context4) {
          while (1) {
            switch (_context4.prev = _context4.next) {
              case 0:
                _context4.next = 2;
                return this._getMultisigENSContract();

              case 2:
                contract = _context4.sent;
                return _context4.abrupt("return", contract.getMultisigName(msigAddress));

              case 4:
              case "end":
                return _context4.stop();
            }
          }
        }, _callee4, this);
      }));

      function getMultisigName(_x) {
        return _getMultisigName.apply(this, arguments);
      }

      return getMultisigName;
    }()
  }, {
    key: "getUserName",
    value: function () {
      var _getUserName = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee5(userAddress) {
        var contract;
        return _regenerator["default"].wrap(function _callee5$(_context5) {
          while (1) {
            switch (_context5.prev = _context5.next) {
              case 0:
                _context5.next = 2;
                return this._getMultisigENSContract();

              case 2:
                contract = _context5.sent;
                return _context5.abrupt("return", contract.getUserName(userAddress));

              case 4:
              case "end":
                return _context5.stop();
            }
          }
        }, _callee5, this);
      }));

      function getUserName(_x2) {
        return _getUserName.apply(this, arguments);
      }

      return getUserName;
    }()
    /**Setters */

  }, {
    key: "setMultisigDeployerAddress",
    value: function () {
      var _setMultisigDeployerAddress = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee6(address) {
        var contract;
        return _regenerator["default"].wrap(function _callee6$(_context6) {
          while (1) {
            switch (_context6.prev = _context6.next) {
              case 0:
                _context6.next = 2;
                return this._getMultisigENSContract();

              case 2:
                contract = _context6.sent;
                return _context6.abrupt("return", contract.functions.setMultisigDeployerAddress(address));

              case 4:
              case "end":
                return _context6.stop();
            }
          }
        }, _callee6, this);
      }));

      function setMultisigDeployerAddress(_x3) {
        return _setMultisigDeployerAddress.apply(this, arguments);
      }

      return setMultisigDeployerAddress;
    }()
  }, {
    key: "setUserName",
    value: function () {
      var _setUserName = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee7(name) {
        var contract;
        return _regenerator["default"].wrap(function _callee7$(_context7) {
          while (1) {
            switch (_context7.prev = _context7.next) {
              case 0:
                _context7.next = 2;
                return this._getMultisigENSContract();

              case 2:
                contract = _context7.sent;
                return _context7.abrupt("return", contract.functions.setUserName(name));

              case 4:
              case "end":
                return _context7.stop();
            }
          }
        }, _callee7, this);
      }));

      function setUserName(_x4) {
        return _setUserName.apply(this, arguments);
      }

      return setUserName;
    }()
    /** Helpers */

  }], [{
    key: "encodeCall",
    value: function encodeCall(functionName, args) {
      return (0, _encodeCall.encode)(artifact, functionName, args);
    }
  }, {
    key: "decodeCall",
    value: function decodeCall(data) {
      return (0, _encodeCall.decode)(artifact, data);
    }
  }]);
  return MultisigENSController;
}();

exports.MultisigENSController = MultisigENSController;
